---
title: "TCGAbiolinks"
author: "yincy"
date: "7/16/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r load packages, message=FALSE, warning=FALSE}
library(TCGAbiolinks)
library(tidyverse)
```


# Introduction  
TCGAbiolinks is able to access The National Cancer Institute (NCI) Genomic Data Commons (GDC) thorough its [GDC Application Programming Interface (API))[https://gdc.cancer.gov/developers/gdc-application-programming-interface-api] to search, download and prepare relevant data for analysis in R.  


# Searching GDC database  
**Differebt sources: Legacy vs Harmonized**  
There are two available sources to download GDC data using TCGAbiolinks:  

- GDC legacy Archive: provides access to an unmodified copy of data that was previously stored in `CGHub` and in the TCGA Data Portal hosted by the TCGA Data Coordinating Center (DCC), in which uses as references GRCh37 (hg19) and GRCh36 (hg18).  

- GDC harmonized database: data available was harmonized against GRCh38 (hg38) using GDC Bioinformatics Pipelines which provides methods to the standardization of biospecimen and clinical data.  


**Understanding the barcode**  
A TCGA barcode is composed of a collection of identifiers. Each specifically identifies a TCGA data element. Refer to the following figure for an illustration of how metadata identifiers comprise a barcode. An aliquot barcode contains the highest number of identifers.  

Example:  

- Aliquot barcode: TCGA-G4-6317-02A-11D-2064-05  
- Participant: TCGA-G4-6317  
- Sample: TCGA-G4-6317-02  

More information at [TCGA barcode](https://docs.gdc.cancer.gov/Encyclopedia/pages/TCGA_Barcode/) 


## Searching arguments  
You can easily search GDC data using the `GDCquery` function.  

Using a summary of filters as used in the TCGA portal, the function works with the following arguments:  

project               | A list of valid project   
----------------------|------------------------------------
data.category         | A valid project (TCGAbiolinks::getProjectSummary(project))  
data.type             | A data type to filter the files to download  
workflow.type         | GDC workflow type  
legacy                | Search in the legacy repository   
access                | Filter by access type. Possible values: controlled, open  
platform              | examples: IlluminaGA_RNASeqV2, IlluminaHiSeq_TotalRNASeqV2  
file.type             | to be used in the legacy database for some platforms, to define which file types to be used.  
barcode               | A list of barcodes to filter the files to download  
experimental.strategy | Filter to experiment strategy. Harmonized: WXS, RNA-seq, miRNA-seq, Genotyping Arry. Legacy: WXS, RNA-seq, miRNA-seq, Genotyping Array, DNA-seq, Methylation array, Protein expression array ...,  
sample.type           | A sample type to filter the files to download  


## Harmonized database examples  
### DNA methylation data: Recurrent tumor samples  
In this example we will access the harmonized database (`legacy = FALSE`) and search for all DNA methylation data for recurrent glioblastoma multiform (GBM) and low grade gliomas (LGG) samples.  
```{r}
query <- GDCquery(project = c("TCGA-GBM", "TCGA-LGG"), 
                  data.category = "DNA Methylation", 
                  platform = c("Illumina Human Methylation 450"), 
                  sample.type = "Recurrent Tumor")

getResults(query)
```



### Samples with DNA methylation and gene expression data  
In this example we will access the harmonized database (`legacy = FALSE`) and search fro all patients with DNA methylation (platform HumanMethylation450k) and gene expression data for Colon Adenocarcinoma tumor (TCGA-COAD).  
```{r}
query.met <- GDCquery(project = "TCGA-COAD", 
                      data.category = "DNA Methylation", 
                      legacy = FALSE, 
                      platform = "Illumina Human Methylation 450")

query.exp <- GDCquery(project = "TCGA-COAD", 
                      data.category = "Transcriptome Profiling", 
                      data.type = "Gene Expression Quantification", 
                      workflow.type = "HTSeq - FPKM-UQ")


# get all patients that have DNA methylation and gene expression  
common.patients <- intersect(substr(getResults(query.met, cols = "cases"), 1, 12), 
                             substr(getResults(query.exp, cols = "cases"), 1, 12))


# only select the first 5 patients  
query.met <- GDCquery(project = "TCGA-COAD", 
                      data.category = "DNA Methylation", 
                      legacy = F, 
                      platform = "Illumina Human Methylation 450", 
                      barcode = common.patients[1:5])

query.exp <- GDCquery(project = "TCGA-COAD", 
                      data.category = "Transcriptome Profiling", 
                      data.type = "Gene Expression Quantification", 
                      barcode = common.patients[1:5], 
                      workflow.type = "HTSeq - FPKM-UQ")
```

```{r}
getResults(query.met)
```
```{r}
getResults(query.exp)
```

### Raw Sequencing Data: Finding the match between file names and barcode for Controlled data  
This example shows how the user can search for breast cancer Raw Sequencing data ("controlled") and verify the name of the files and the barcodes associated with it.  
```{r}
query <- GDCquery(project = "TCGA-BRCA", 
                  data.category = "Sequencing Reads", 
                  sample.type = "Primary Tumor")

# only first 100 to make render faster  
getResults(query, rows = 1:100, cols = c("file_name", "cases"))
```

## Legacy archive example  
### DNA methylation   
#### Array-based assays  
This example shows how the user can search for glioblastoma multiform (GBM) and low grade gliomas (LGG) DNA methylation data for platform Illumina Human Methylation 450 and Illumina Human Methylation 27.  
```{r}
query <- GDCquery(project = c("TCGA-GBM", "TCGA-LGG"), 
                  legacy = T, 
                  platform = c("Illumina Human Methylation 450", "Illumina Human Methylation 27"), 
                  data.category = "DNA methylation")

getResults(query, rows = 1:100) %>% tibble::as_tibble()
```


### whole-genome bisulfite sequencing (WGBS)  
```{r}
query <- GDCquery(project = "TCGA-LUAD", 
                  legacy = T, 
                  data.category = "DNA methylation", 
                  data.type = "Methylation percentage", 
                  experimental.strategy = "Bisulfite-Seq")


#VCF - controlled data 
query <- GDCquery(project = "TCGA-LUAD", 
                  legacy = T, 
                  data.category = "DNA methylation", 
                  data.type = "Bisulfite sequence alignment", 
                  experimental.strategy = "Bisulfite-Seq")


# WGBS BAM files - controlled data  
query <- GDCquery(project = "TCGA-LUAD", 
                  legacy = T, 
                  data.type = "Aligned reads", 
                  data.category = "Raw sequencing data", 
                  experimental.strategy = "Bisulfite-Seq")

getResults(query)
```


### Gene expression  
This example shows how the user can search for glioblastoma multiform (GBM) gene expression data with the normalized results for expression of a gene. For more information about file.types check [GDC TCGA file types](https://gdc.cancer.gov/resources-tcga-users/legacy-archive-tcga-tag-descriptions).  
```{r}
query.exp.hg19 <- GDCquery(project = "TCGA-GBM", 
                           data.category = "Gene expression", 
                           data.type = "Gene expression quantification", 
                           platform = "Illumina HiSeq", 
                           file.type = "normalized_results", 
                           experimental.strategy = "RNA-Seq", 
                           barcode = c("TCGA-14-0736-02A-01R-2005-01", "TCGA-06-0211-02A-02R-2005-01"), 
                           legacy = T)

getResults(query.exp.hg19)
```


### Get Manifest file  
If you want to get the manifest file from the query object you can use the function `getManifest`.  
```{r}
getManifest(query = query.exp.hg19, save = F)
```


## ATAC-seq data  
For the moment, ATAC-seq data is available at the [GDC publication page](https://gdc.cancer.gov/about-data/publications/ATACseq-AWG). Also, for more details, you can check an ATAC-seq workshop at http://rpubs.com/tiagochst/atac_seq_workshop  

```{r, message=FALSE, warning=FALSE}
GDCquery_ATAC_seq() %>% getResults() %>% as_tibble()
```

You can use the function `GDCquery_ATAC_seq` filter the manifest table and use `GDCdownload` to save the data locally.  
```{r}
query <- GDCquery_ATAC_seq(file.type = "rds")
GDCdownload(query = query, method = "client")
```

```{r}
query <- GDCquery_ATAC_seq(file.type = "bigWigs")
GDCdownload(query = query, method = "client")
```


## Summary of available files per patient  
Retrieve the number of files under each data_category _ data_type + experiemtnal_strategy _ platform. Almost like https://portal.gdc.cancer.gov/exploration  
```{r}
tab <- getSampleFilesSummary(project = "TCGA-ACC")

tab %>% as_tibble()
```


# Downloading and preparing files for analysis  
**TCGAbiolinks** has provided a few functions to download and prepare data from GDC for analysis. This section starts by explaining the different downloads methods and the SummarizedExperiment object which is the default data structure used in TCGAbiolinks.  

## Downloading and preparing data for analysis 
**Data downloading: Methods differences**    

- client: this method create a MANIFEST file and download the data using [GDC Data Transfer Tool](https://gdc.cancer.gov/access-data/gdc-data-transfer-tool) this method is more reliable but it might be slower compared to the api method.  
- api: this methods used the [GDC Application Programming Interface (API)](https://gdc.cancer.gov/developers/gdc-application-programming-interface-api) to download the data. This will create a MANIFEST file and the data downloaded will be compressed into a tar.gz file. If the size and the number of the files are too big this tar.gz will be too big which might have high probability of download failure. To solve that we created the `files.per.chunk` argument which will split the files into small chunks, for example, if chunks.per.download is equal to 10 we will download only 10 files inside each tar.gz.   

**Data prepared:SummarizedExperiment object**  
A `SummarizedExperiment` object has three main matrices that can be accessed using the `SummarizedExperiemnt` package:  

- Sample matrix information is accessed via `colData(data`: stores sample information. TCGAbiolinks will add indexed clinicl data and subtype information from marker TCGA papers.  
- Assay matrix information is accessed via `assay(data`: stores molecula data   
- Feature matrix information (gene information) is accessed via `rowRanges(data)`: stores metadata about the features, including their genomic ranges.  

**Summarized Experiment: annotation information**  
When using thefunction `GDCprepare` there is an argument called `SummarizedExperiment` which defines the output type a Summarized Experiment (default option) or a data frame. To create a summarized Experiment object we annotate the data with genomic positions with last patch release version of the genome available. For legacy data (data aligned to hg19) TCGAbiolinks is using GRCh37.p13 and for harmonized data (data aligned to hg38) now it is using GRCh38.p7.  








