---
title: "TCGAbiolinks"
author: "yincy"
date: "7/16/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r load packages, message=FALSE, warning=FALSE}
library(TCGAbiolinks)
library(tidyverse)
```


# Introduction  
TCGAbiolinks is able to access The National Cancer Institute (NCI) Genomic Data Commons (GDC) thorough its [GDC Application Programming Interface (API))[https://gdc.cancer.gov/developers/gdc-application-programming-interface-api] to search, download and prepare relevant data for analysis in R.  


# Searching GDC database  
**Differebt sources: Legacy vs Harmonized**  
There are two available sources to download GDC data using TCGAbiolinks:  

- GDC legacy Archive: provides access to an unmodified copy of data that was previously stored in `CGHub` and in the TCGA Data Portal hosted by the TCGA Data Coordinating Center (DCC), in which uses as references GRCh37 (hg19) and GRCh36 (hg18).  

- GDC harmonized database: data available was harmonized against GRCh38 (hg38) using GDC Bioinformatics Pipelines which provides methods to the standardization of biospecimen and clinical data.  


**Understanding the barcode**  
A TCGA barcode is composed of a collection of identifiers. Each specifically identifies a TCGA data element. Refer to the following figure for an illustration of how metadata identifiers comprise a barcode. An aliquot barcode contains the highest number of identifers.  

Example:  

- Aliquot barcode: TCGA-G4-6317-02A-11D-2064-05  
- Participant: TCGA-G4-6317  
- Sample: TCGA-G4-6317-02  

More information at [TCGA barcode](https://docs.gdc.cancer.gov/Encyclopedia/pages/TCGA_Barcode/) 


## Searching arguments  
You can easily search GDC data using the `GDCquery` function.  

Using a summary of filters as used in the TCGA portal, the function works with the following arguments:  

project               | A list of valid project   
----------------------|------------------------------------
data.category         | A valid project (TCGAbiolinks::getProjectSummary(project))  
data.type             | A data type to filter the files to download  
workflow.type         | GDC workflow type  
legacy                | Search in the legacy repository   
access                | Filter by access type. Possible values: controlled, open  
platform              | examples: IlluminaGA_RNASeqV2, IlluminaHiSeq_TotalRNASeqV2  
file.type             | to be used in the legacy database for some platforms, to define which file types to be used.  
barcode               | A list of barcodes to filter the files to download  
experimental.strategy | Filter to experiment strategy. Harmonized: WXS, RNA-seq, miRNA-seq, Genotyping Arry. Legacy: WXS, RNA-seq, miRNA-seq, Genotyping Array, DNA-seq, Methylation array, Protein expression array ...,  
sample.type           | A sample type to filter the files to download  


## Harmonized database examples  
### DNA methylation data: Recurrent tumor samples  
In this example we will access the harmonized database (`legacy = FALSE`) and search for all DNA methylation data for recurrent glioblastoma multiform (GBM) and low grade gliomas (LGG) samples.  
```{r}
query <- GDCquery(project = c("TCGA-GBM", "TCGA-LGG"), 
                  data.category = "DNA Methylation", 
                  platform = c("Illumina Human Methylation 450"), 
                  sample.type = "Recurrent Tumor")

getResults(query)
```



### Samples with DNA methylation and gene expression data  
In this example we will access the harmonized database (`legacy = FALSE`) and search fro all patients with DNA methylation (platform HumanMethylation450k) and gene expression data for Colon Adenocarcinoma tumor (TCGA-COAD).  
```{r}
query.met <- GDCquery(project = "TCGA-COAD", 
                      data.category = "DNA Methylation", 
                      legacy = FALSE, 
                      platform = "Illumina Human Methylation 450")

query.exp <- GDCquery(project = "TCGA-COAD", 
                      data.category = "Transcriptome Profiling", 
                      data.type = "Gene Expression Quantification", 
                      workflow.type = "HTSeq - FPKM-UQ")


# get all patients that have DNA methylation and gene expression  
common.patients <- intersect(substr(getResults(query.met, cols = "cases"), 1, 12), 
                             substr(getResults(query.exp, cols = "cases"), 1, 12))


# only select the first 5 patients  
query.met <- GDCquery(project = "TCGA-COAD", 
                      data.category = "DNA Methylation", 
                      legacy = F, 
                      platform = "Illumina Human Methylation 450", 
                      barcode = common.patients[1:5])

query.exp <- GDCquery(project = "TCGA-COAD", 
                      data.category = "Transcriptome Profiling", 
                      data.type = "Gene Expression Quantification", 
                      barcode = common.patients[1:5], 
                      workflow.type = "HTSeq - FPKM-UQ")
```

```{r}
getResults(query.met)
```
```{r}
getResults(query.exp)
```

### Raw Sequencing Data: Finding the match between file names and barcode for Controlled data  
This example shows how the user can search for breast cancer Raw Sequencing data ("controlled") and verify the name of the files and the barcodes associated with it.  
```{r}
query <- GDCquery(project = "TCGA-BRCA", 
                  data.category = "Sequencing Reads", 
                  sample.type = "Primary Tumor")

# only first 100 to make render faster  
getResults(query, rows = 1:100, cols = c("file_name", "cases"))
```

## Legacy archive example  
### DNA methylation   
#### Array-based assays  
This example shows how the user can search for glioblastoma multiform (GBM) and low grade gliomas (LGG) DNA methylation data for platform Illumina Human Methylation 450 and Illumina Human Methylation 27.  
```{r}
query <- GDCquery(project = c("TCGA-GBM", "TCGA-LGG"), 
                  legacy = T, 
                  platform = c("Illumina Human Methylation 450", "Illumina Human Methylation 27"), 
                  data.category = "DNA methylation")

getResults(query, rows = 1:100) %>% tibble::as_tibble()
```


### whole-genome bisulfite sequencing (WGBS)  
```{r}
query <- GDCquery(project = "TCGA-LUAD", 
                  legacy = T, 
                  data.category = "DNA methylation", 
                  data.type = "Methylation percentage", 
                  experimental.strategy = "Bisulfite-Seq")


#VCF - controlled data 
query <- GDCquery(project = "TCGA-LUAD", 
                  legacy = T, 
                  data.category = "DNA methylation", 
                  data.type = "Bisulfite sequence alignment", 
                  experimental.strategy = "Bisulfite-Seq")


# WGBS BAM files - controlled data  
query <- GDCquery(project = "TCGA-LUAD", 
                  legacy = T, 
                  data.type = "Aligned reads", 
                  data.category = "Raw sequencing data", 
                  experimental.strategy = "Bisulfite-Seq")

getResults(query)
```


### Gene expression  
This example shows how the user can search for glioblastoma multiform (GBM) gene expression data with the normalized results for expression of a gene. For more information about file.types check [GDC TCGA file types](https://gdc.cancer.gov/resources-tcga-users/legacy-archive-tcga-tag-descriptions).  
```{r}
query.exp.hg19 <- GDCquery(project = "TCGA-GBM", 
                           data.category = "Gene expression", 
                           data.type = "Gene expression quantification", 
                           platform = "Illumina HiSeq", 
                           file.type = "normalized_results", 
                           experimental.strategy = "RNA-Seq", 
                           barcode = c("TCGA-14-0736-02A-01R-2005-01", "TCGA-06-0211-02A-02R-2005-01"), 
                           legacy = T)

getResults(query.exp.hg19)
```


### Get Manifest file  
If you want to get the manifest file from the query object you can use the function `getManifest`.  
```{r}
getManifest(query = query.exp.hg19, save = F)
```


## ATAC-seq data  
For the moment, ATAC-seq data is available at the [GDC publication page](https://gdc.cancer.gov/about-data/publications/ATACseq-AWG). Also, for more details, you can check an ATAC-seq workshop at http://rpubs.com/tiagochst/atac_seq_workshop  

```{r, message=FALSE, warning=FALSE}
GDCquery_ATAC_seq() %>% getResults() %>% as_tibble()
```

You can use the function `GDCquery_ATAC_seq` filter the manifest table and use `GDCdownload` to save the data locally.  
```{r}
query <- GDCquery_ATAC_seq(file.type = "rds")
GDCdownload(query = query, method = "client")
```

```{r}
query <- GDCquery_ATAC_seq(file.type = "bigWigs")
GDCdownload(query = query, method = "client")
```


## Summary of available files per patient  
Retrieve the number of files under each data_category _ data_type + experiemtnal_strategy _ platform. Almost like https://portal.gdc.cancer.gov/exploration  
```{r}
tab <- getSampleFilesSummary(project = "TCGA-ACC")

tab %>% as_tibble()
```


# Downloading and preparing files for analysis  
**TCGAbiolinks** has provided a few functions to download and prepare data from GDC for analysis. This section starts by explaining the different downloads methods and the SummarizedExperiment object which is the default data structure used in TCGAbiolinks.  

## Downloading and preparing data for analysis 
**Data downloading: Methods differences**    

- client: this method create a MANIFEST file and download the data using [GDC Data Transfer Tool](https://gdc.cancer.gov/access-data/gdc-data-transfer-tool) this method is more reliable but it might be slower compared to the api method.  
- api: this methods used the [GDC Application Programming Interface (API)](https://gdc.cancer.gov/developers/gdc-application-programming-interface-api) to download the data. This will create a MANIFEST file and the data downloaded will be compressed into a tar.gz file. If the size and the number of the files are too big this tar.gz will be too big which might have high probability of download failure. To solve that we created the `files.per.chunk` argument which will split the files into small chunks, for example, if chunks.per.download is equal to 10 we will download only 10 files inside each tar.gz.   

**Data prepared:SummarizedExperiment object**  
A `SummarizedExperiment` object has three main matrices that can be accessed using the `SummarizedExperiemnt` package:  

- Sample matrix information is accessed via `colData(data`: stores sample information. TCGAbiolinks will add indexed clinicl data and subtype information from marker TCGA papers.  
- Assay matrix information is accessed via `assay(data`: stores molecula data   
- Feature matrix information (gene information) is accessed via `rowRanges(data)`: stores metadata about the features, including their genomic ranges.  

**Summarized Experiment: annotation information**  
When using thefunction `GDCprepare` there is an argument called `SummarizedExperiment` which defines the output type a Summarized Experiment (default option) or a data frame. To create a summarized Experiment object we annotate the data with genomic positions with last patch release version of the genome available. For legacy data (data aligned to hg19) TCGAbiolinks is using GRCh37.p13 and for harmonized data (data aligned to hg38) now it is using GRCh38.p7.  

Unfortunately, some of the updates change/remove gene symbols, change coordinates, etc. Which might introduce some loss of data. For example, if the gene was removed we cannot map it anymore and that information will be lost in the `SummarizedExperiment`.  

If you set `SummarizedExperiment` to FALSE, you will get the data unmodified just as they are in the files and add your own annotation.  

Also, there are no updated for DNA methylation data. But the last metadata available can be found here: http://zwdzwd.github.io/InfiniumAnnotation  


## Arguments  
`GDCdownload`  
**Argument**|**Description**  
------------|----------------------
query|A query for GDCquery function
token.file|Token file to download controlled data (only for method = "client")
method|Uses the API (POST method) or gdc client tool. Options "api", "client". API is faster, but the data might get corrupted in the download, and it might need to be executed again
directory|Directory/Folder where the data was downloaded. Default: GDCdata
files.per.chunk|This will make the API method only download n (files.per.chunk) files at a time. This may reduce the download problems when the data size is too large. Expected a integer number (example files.per.chunk=6)


`GDCprepare`  
**Argument**|**Description**
------------|------------------------
query|A query for GDCquery function
save|Save result as RData object?
save.filename|Name of the file to be save if empty an automatic will be created
directory|Directory/Folder where the data was downloaded. Default:GDCdata
summarizedExperiment|Create a summarizedExperiment?Default TRUE
remove.files.prepared|Remove the file read? Default: FALSE This argument will be considered only if save argument is set to true.
add.gistic2.mut|If a list of genes (gene symbol) is given, columns with gistic2 results from GDAC firehouse (hg19) and a column indicating if there is or not mutation in that gene (hg38) (TRUE or FALSE - use the MAF file for more information) will be added to the sample matrix in the summarized Experiment object.
mut.pipline|If add.gistic2.mut is not NULL this field will be taken in consideration. Four separate variant calling piplines are implemented for GDC data harmonization. Options: muse, varscan2, somaticsniper, Mutect2. For more information:https://gdc-docs.nci.nih.gov/Data/Bioinformatics_Pipelines/DNA_Seq_Variant_Calling_Pipeline/
mutant_variant_classification|List of mutant_variant_classification that will be consider a smple mutant or not. Default:"Frame_Shift_Del", "Frame_Shift_ins", "Misssense_Mutation", "Nonsense_Mutation", "Splice_Site", "In_Frame_Ins", "Translation_Start_Site", "Nonstop_Mutation"  


## Search and download data from legacy database using GDC api method  
In this example we will download gene expression data from legacy database (data aligned against genome of reference hg19) using GDC api method and we will show object data and metadata.  
```{r}
query <- GDCquery(project = "TCGA-GBM", 
                  data.type = "Gene expression quantification", 
                  data.category = "Gene expression", 
                  platform = "Illumina HiSeq", 
                  file.type = "normalized_result", 
                  experimental.strategy = "RNA-Seq", 
                  barcode = c("TCGA-14-0736-02A-01R-2005-01", "TCGA-06-0211-02A-02R-2005-01"), 
                  legacy = T)

GDCdownload(query = query, 
            method = "api", 
            files.per.chunk = 10)
```

```{r}
data <- GDCprepare(query = query)
colData(data) %>% as_tibble()
```

```{r}
assay(data) %>% as.data.frame()
```

```{r}
rowRanges(data) 
```


## Search and download data for two samples from database  
In this example we will download gene expression quantification from harmonized database (data aligned against genome of reference hg38). Also, it shows the object data and metadata.  
```{r}
# gene expression aligned against hg38
query <- GDCquery(project = "TCGA-GBM", 
                  data.category = "Transcriptome Profiling", 
                  data.type = "Gene Expression Quantification", 
                  workflow.type = "HTSeq - FPKM-UQ", 
                  barcode = c("TCGA-14-0736-02A-01R-2005-01", "TCGA-06-0211-02A-02R-2005-01"))

GDCdownload(query = query)
data <- GDCprepare(query = query)
```


```{r}
colData(data) %>% as.data.frame()
```

```{r}
rowRanges(data)
```

```{r}
assay(data) %>% as.data.frame()
```



## `GDCprepare`: Outputs  
This function is still under development, it is not working for all cases. See the tables below with thte status. Examples of query, download, prepare can be found in this [gist](https://gist.github.com/tiagochst/a701bad3fa3800ade7063760755e0aad)  


### Harmonized data  
**Data.category**|**Data.type**|**Workflow Type**|**Status**
-----------------|-------------|-----------------|-----------------
Transcriptome Profiling|Gene Expression Quantification|THSeq-Counts|Data frame or SE (losing 5% of information when mapping to genomic regions)
|||HTSeq-FPKM-UQ|Returning only a (losing 5% of information when mapping to genomic regions)
|||HTSeq-FPKM|Returning only a (losing 5% of information when mappping to genomic regions)
||Isoform Expression Quantification|Not needed||
|miRNA Expression Quantification|Not needed|Returning only a dataframe for the moment
Copy number variation|Copy Number Segment|Returning only a dataframe for the moment
||Masked Copy NUmber Segment|Returning only a dataframe for the moment
||Gene Level Copy Number Scores|Returning only a dataframe for the moment
Simple Nucleotide Variation|||
Raw Sequencing Data||
Biospecimen|Slide Image||
Biospecimen|Biospecimen Supplement||
Clinical|||


## **Legacy data**  
**Data.category**|**Data.type**|**Platform**|**file.type**|**Status**
-----------------|-------------|------------|-------------|--------------------
Transcriptome Profiling||||
Copy number variation|-|Affymetrix SNP Array 6.0|nocnv_hg18.seg|Working
||-|Affymetrix SNP Array 6.0|hg18.seg|Working
||-|Affymetrix SNP Array 6.0|nocnv_hg19.seg|Working
||-|Affymetrix SNP Array 6.0|hg19.seg|Working
Simple Nucleotide Variation| Simple somatic mutation||
Raw Sequencing data||||
Biospecimen||||
Clinical||||
Protein expression|MDA RPPA Core|-|Working
Gene Expression|Gene expression quantification|Illumina HiSeq|normalized_results|Working
|||Illumina HiSeq|results|Working
|||HT_HG_U133A|-|
|||AgilentG4502A_07_2|-|Data frame only
|||AgilentG4502A_07_1|-|Data frame only
|||HuEx-1_0-st-v2|FIRMA.txt|Not Preparing
|||gene.txt|Not Preparing
|Isoform expression quantification||
|miRNA gene quantification||
|Exon junction quantification||
|Exon quantification||
|miRNA isoform quantification||
DNA methylation||Illumina Human Methylation 450|Not used|Working
|||Illumina Human Methylation 27|Not used|Working
|||Illumina DNA Methylation OMA003 CPI|Not used|Working
|||Illumina DNA Methylation OMA002 CPI|Not used|Working
|||Illumina HiSeq||Not Working
Raw Microarray Data||||
Structure Rearrangement||||
Other||||


## Examples 
### Legacy archive  
#### DNA methylation:Get all TCGA IDAT files  
Example to idat files from TCGA projects  
```{r}
projects <- getGDCprojects()[, "project_id", drop = T] %>% unique()
projects <- projects[grepl("TCGA", projects)]
match.file.cases.all <- NULL

for(proj in projects){
    print(proj)
    query <- GDCquery(project = proj, 
                      data.category = "Raw microarray data", 
                      data.type = "Raw intensities", 
                      experimental.strategy = "Methylation array", 
                      legacy = T, 
                      file.type = ".idat", 
                      platform = "Illumina Human Methylation 450")
    
    match.file.cases <- getResults(query, cols = c("cases", "file_name"))
    match.file.cases$project <- proj
    match.file.cases.all <- rbind(match.file.cases.all, match.file.cases)
    tryCatch(GDCdownload(query, method = "api", files.per.chunk = 20), 
             error = function(e)GDCdownload(query, method = "client"))
}
```


#### DNAmethylation: aligned against hg19  
```{r}
query_meth.hg19 <- GDCquery(project = "TCGA-LGG", 
                            data.category = "DNA methylation", 
                            platform = "Illumina Human Methylation 450", 
                            barcode = c("TCGA-HT-8111-01A-11D-2399-05","TCGA-HT-A5R5-01A-11D-A28N-05"), 
                            legacy = T)

GDCdownload(query = query_meth.hg19)
data.hg19 <- GDCprepare(query = query_meth.hg19)
rowRanges(data.hg19)
assay(data.hg19) %>% as.data.frame()
colData(data.hg19) %>% as.data.frame()
```

#### Protein expression  
```{r}
query <- GDCquery(project = "TCGA-GBM", 
                  data.category = "Protein expression", 
                  legacy = T, 
                  barcode = c("TCGA-OX-A56R-01A-21-A44T-20","TCGA-08-0357-01A-21-1898-20"))

GDCdownload(query = query, 
            save = T, 
            save.filename = "gbmProteinExpression.rda", 
            remove.files.prepared = TRUE)
```


# Clinical Data  
TCGAbiolinks has provided a few functions to search, download and parse clinical data. This section starts by explaining the different sources for clinical information in GDC, followed by the necessary function to access these sources.  

## Usefull information  
**Different sources**  
In GDC database the clinical data can be retrieved from different sources:  

- indexed clinical: a refined clinical data that is created using the XML files.  
- XML files: original source of the data  
- BCR Biotab: tsv files parsed from XML files  

There are two main difference between the indexed clinical and XML files:  

- XML has more information:radiation, drugs information, follow up information, etc. So the indexed one is only subset of the XML files.  
- The indexed data contains the updated data with the follow up information. For example: if the patient is alive in the first time clinical data was collect and the in the next follow-up he is dead, the indexed data will show dead. The XML will have two fields, one for the first time saying he is alive (in the clinical part) and the follow-up saying he is dead.  


Other useful clinical information available are:  
- Tissue slide image  
- Pathology report-Slide image  


## BCR Biotab  
### Clinical  
In this example we will fetch clinical data from BCR Biotab files.  
```{r}
query <- GDCquery(project = "TCGA-ACC", 
                  data.category = "Clinical", 
                  data.type = "Clinical Supplement", 
                  data.format = "BCR Biotab")

GDCdownload(query = query)
clinical.BCRtab.all <- GDCprepare(query)
clinical.BCRtab.all %>% names()


query <- GDCquery(project = "TCGA-ACC", 
                  data.category = "Clinical", 
                  data.type = "Clinical Supplement", 
                  data.format = "BCR Biotab", 
                  file.type = "radiation")

GDCdownload(query = query)
clinical.BCRtab.raditation <- GDCprepare(query)
clinical.BCRtab.raditation %>% class()
clinical.BCRtab.raditation$clinical_radiation_acc
```

```{r}
query <- GDCquery(project = "TCGA-BRCA", 
                  data.category = "Clinical", 
                  data.type = "Clinical Supplement", 
                  data.format = "BCR Biotab")

GDCdownload(query = query)
clinical.BCRtab.all <- GDCprepare(query)
```

```{r}
clinical.BCRtab.all %>% names()
```

```{r}
clinical.BCRtab.all %>% class()
clinical.BCRtab.all$clinical_patient_brca %>% colnames() %>% tibble::as_tibble()
```

ER status count 
```{r}
clinical.BCRtab.all %>% .$clinical_patient_brca %>% .$er_status_by_ihc %>% table()
```

ER content  
```{r}
er.cols <- grep("^er", colnames(clinical.BCRtab.all$clinical_patient_brca))
clinical.BCRtab.all$clinical_patient_brca[, c(2, er.cols)] %>% 
    DT::datatable(options = list(scrollX = TRUE))
```

```{r}
clinical.BCRtab.all$clinical_patient_brca %>% 
    head() %>% 
    DT::datatable(options = list(scrollX = T, keys = TRUE))
```

### Biospecimen  
Biospecimen BCR Biotab  
```{r}
query.biospeciemn <- GDCquery(project = "TCGA-BRCA", 
                              data.category = "Biospecimen", 
                              data.type = "Biospecimen Supplement", 
                              data.format = "BCR Biotab")

GDCdownload(query = query.biospeciemn)
biospecimen.BCRtab.all <- GDCprepare(query = query.biospeciemn)
```

All available tables  
```{r}
biospecimen.BCRtab.all %>% names()
biospecimen.BCRtab.all$biospecimen_aliquot_brca 
```

### Clinical indexed data  
```{r}
clinical <- GDCquery_clinic(project = "TCGA-LUAD", type = "clinical")

clinical %>% 
    head() %>% 
    DT::datatable(options = list(scrollX = TRUE))


clinical <- GDCquery_clinic(project = "CPTAC-2", type = "clinical")
```


### XML clinical data  
The process to get data directly from the XML are:  

1. Use `GDCquery` and `GDCdownload` functions to search/download either biospecimen or clinical XML files.  
2. Use `GDCprepare_clinic` function to parse the XML files.  

The relation between one patient and other clinical information are 1:n, one patient could have several raditation treatments. For that reason, we only give the option to parse individual tables (only drug information, only radiation information, ...) The selection of the table is done by the argument `clinical.info`.  

*clinical.info options to parse information for each data category*  

**data.category**  | **clinical.info**
-------------------|-----------------------------------
Clinical| drug
Clinical| admin
Clinical| follow_up
Clinical| radiation
Clinical| patient
Clinical| stage_event
Clinical| new_tumor_event
Biospecimen| sample
Biospecimen| bio_patient
Biospecimen| analyte 
Biospecimen| aliquot
Biospecimen| protocol
Biospecimen| portion
Biospecimen| slide
Other| msi


```{r}
query <- GDCquery(project = "TCGA-COAD", 
                  data.category = "Clinical", 
                  file.type = "xml", 
                  barcode = c("TCGA-RU-A8FL", "TCGA-AA-3972"))

GDCdownload(query)
clinical <- GDCprepare_clinic(query, clinical.info = "patient")

clinical %>% 
    DT::datatable(options = list(scrollX = T))
```

```{r}
clinical.radiation <- GDCprepare_clinic(query = query, clinical.info = "radiation")

clinical.radiation %>% 
    DT::datatable(options = list(scrollX = T, 
                                 rowname = F))
```

```{r}
clinical.admin <- GDCprepare_clinic(query = query, clinical.info = "admin")

clinical.admin %>% 
    DT::datatable(options = list(scrollX = T))
```


# Microsatellite data  
MSI-Mono-Dinucleotide Assay is performed to test a panel of four mononucleotide repeat loci (polyadenine tracts BAT25, BAT26, BAT40, and transforming growth factor receptor type II) and three dinucleotide repeat loci (CA repeats in D2S123, D5S346, and D17S250). Two additional pentanucleotide loci (Penta D and Penta E) are included in this assay to evaluate sample identity. Multiplex fluorecent-labeld PCR and capillary electrophoresis were used to identify MSI if a variation in the number of microsatellite repeats was detected between tumor and matched non-neoplastic tissue or mononuclear blood cells. Equivocal or failed markers were altered nad high level MSI (MSI-H) if greater than 40% of markers were altered.  


Level3 data is included in BCR clinical-based submissions and can be downloaded as follows:  

```{r}
quey <- GDCquery(project = "TCGA-COAD", 
                 data.category = "Other", 
                 legacy = T, 
                 access = "open", 
                 data.type = "Auxiliary test", 
                 barcode = c("TCGA-AD-A5EJ", "TCGA-DM-A0X9"))

GDCdownload(quey)
msi_results <- GDCprepare_clinic(quey, "msi")

msi_results %>% 
    DT::datatable(options = list(scrollX = T))
```


## Tissue slide image (SVS format)  
Tissue slide image files from legacy database  
```{r}
query.legacy <- GDCquery(project = "TCGA-COAD", 
                         data.category = "Clinical", 
                         data.type = "Tissue slide image", 
                         legacy = T, 
                         barcode = c("TCGA-RU-A8FL", "TCGA-AA-3972"))
```

Tissue slide image files from harmonized database  
```{r}
query.harmonized <- GDCquery(project = "TCGA-OV", 
                             data.category = "Biospecimen", 
                             data.type = "Slide Image")

query.harmonized %>% 
    getResults() %>% 
    DT::datatable()
```


## Diagnostic Slide (SVS format)  
Pathology report from harmonized protal  
```{r}
query.harmonized <- GDCquery(project = "TCGA-COAD", 
                             data.category = "Biospecimen", 
                             data.type = "Slide Image", 
                             experimental.strategy = "Diagnostic Slide", 
                             barcode = c("TCGA-RU-A8FL", "TCGA-AA-3972"))

query.harmonized %>% 
    getResults() %>% 
    head() %>% 
    DT::datatable(options = list(scrollX = T))
```


## Legacy archive files  
The clinical data types available in legacy database are:  

- Biospecimen data (Biotab format)  
- Tissue slide image (SVS format)  
- Clinical Supplement (XML format)
- Pathology report (PDF)
- Clinical data (Biotab format)


### Pathology report (PDF)  
pathology report from legacy portal  
```{r}
query.legacy <- GDCquery(project = "TCGA-COAD", 
                         data.category = "Clinical", 
                         data.type = "Pathology report", 
                         legacy = TRUE, 
                         barcode = c("TCGA-RU-A8FL", "TCGA-AA-3972"))

query.legacy %>% 
    getResults() %>% 
    DT::datatable(options = list(scrollX = T))
```


### Tissue slide image (SVS format)  
tissue slide image  
```{r}
query <- GDCquery(project = "TCGA-COAD", 
                  data.category = "Clinical", 
                  data.type = "Tissue slide image", 
                  legacy = T, 
                  barcode = c("TCGA-RU-A8FL", "TCGA-AA-3972"))

query %>% 
    getResults() %>% 
    DT::datatable(options = list(scrollX = T))
```


### Clinical Supplement (XML format)  
clinical supplemetn  
```{r}
query <- GDCquery(project = "TCGA-COAD", 
                  data.category = "Clinical", 
                  data.type = "Clinical Supplement", 
                  legacy = T,
                  barcode = c("TCGA-RU-A8FL", "TCGA-AA-3972"))

query %>% 
    getResults() %>% 
    DT::datatable(options = list(scrollX = T))
```


## Clinical data (Biotab format)  
Clinical data  
```{r}
query <- GDCquery(project = "TCGA-COAD", 
                  data.category = "Clinical", 
                  data.type = "Clinical data", 
                  legacy = T, 
                  file.type = "txt")

query %>% 
    getResults() %>% 
    DT::datatable(options = list(scrollX = T))
```

```{r}
GDCdownload(query = query)

clinical.biotab <- GDCprepare(query = query)

clinical.biotab %>% names()
```

```{r}
clinical.biotab$clinical_radiation_coad %>% as_tibble()
```


### Filter Functions  
Also, some functions to work with clinical data are provided.  

For example, the function `TCGAquery_SampleTypes` will filter barcode based on a type the argument typesample.   

An example of the function is below  

```{r}
bar <- c("TCGA-G9-6378-02A-11R-1789-07", "TCGA-CH-5767-04A-11R-1789-07",  
         "TCGA-G9-6332-60A-11R-1789-07", "TCGA-G9-6336-01A-11R-1789-07",
         "TCGA-G9-6336-11A-11R-1789-07", "TCGA-G9-7336-11A-11R-1789-07",
         "TCGA-G9-7336-04A-11R-1789-07", "TCGA-G9-7336-14A-11R-1789-07",
         "TCGA-G9-7036-04A-11R-1789-07", "TCGA-G9-7036-02A-11R-1789-07",
         "TCGA-G9-7036-11A-11R-1789-07", "TCGA-G9-7036-03A-11R-1789-07",
         "TCGA-G9-7036-10A-11R-1789-07", "TCGA-BH-A1ES-10A-11R-1789-07",
         "TCGA-BH-A1F0-10A-11R-1789-07", "TCGA-BH-A0BZ-02A-11R-1789-07",
         "TCGA-B6-A0WY-04A-11R-1789-07", "TCGA-BH-A1FG-04A-11R-1789-08",
         "TCGA-D8-A1JS-04A-11R-2089-08", "TCGA-AN-A0FN-11A-11R-8789-08",
         "TCGA-AR-A2LQ-12A-11R-8799-08", "TCGA-AR-A2LH-03A-11R-1789-07",
         "TCGA-BH-A1F8-04A-11R-5789-07", "TCGA-AR-A24T-04A-55R-1789-07",
         "TCGA-AO-A0J5-05A-11R-1789-07", "TCGA-BH-A0B4-11A-12R-1789-07",
         "TCGA-B6-A1KN-60A-13R-1789-07", "TCGA-AO-A0J5-01A-11R-1789-07",
         "TCGA-AO-A0J5-01A-11R-1789-07", "TCGA-G9-6336-11A-11R-1789-07",
         "TCGA-G9-6380-11A-11R-1789-07", "TCGA-G9-6380-01A-11R-1789-07",
         "TCGA-G9-6340-01A-11R-1789-07", "TCGA-G9-6340-11A-11R-1789-07")

s <- TCGAquery_SampleTypes(barcode = bar, 
                           typesample = "TP")

s2 <- TCGAquery_SampleTypes(barcode = bar, 
                            typesample = "NB")

# retrieve multiple tissue types NOT from the same patients  
ss <- TCGAquery_SampleTypes(barcode = bar, 
                            typesample = "NB")


# retrieve multiple tissue types FROM the same patients  
ss1 <- TCGAquery_SampleTypes(barcode = bar, 
                             typesample = c("NT", "TP"))
```


## Other useful code  
To get all the information fro TCGA samples you can use the script below  

```{r}
# this code will get all clinical indexed data from TCGA 
library(data.table)
library(dplyr)
library(regexPipes)


getGDCprojects() %>% .[, "project_id"] %>% 
    grep(pattern = "^TCGA", value = T) %>% 
    sort() %>% 
    plyr::alply(1, GDCquery_clinic, .progress = "text") %>% 
    plyr::rbind.fill()
```



# Mutation data 
## Search and Download 
**TCGAbiolinks** has provided a few functions to download mutation data from GDC. There are two options to download the data:  

1. Use `GDCquery_Maf` which will download MAF aligned aganist hg38  
2. Use `GDCquery`, `GDCdownload` and `GDCprepare` to download MAF aligned against hg19  
3. Use `getMC3MAF()` to download MC3 MAF from https://gdc.cancer.gov/about-data/publications/mc3-2017  

## Mutation data (hg38)  
This example will download MAF (mutation annotation files) for variant calling pipeline muse. Piplines options are: `muse`, `varscan2`, `somaticsniper`, `mutect`. For more information please access [GDC docs](https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/DNA_Seq_Variant_Calling_Pipeline/)  

```{r}
maf <- GDCquery_Maf(tumor = "CHOL", pipelines = "muse")

# only first 50 to make render faster
DT::datatable(maf[1:20, ], 
              filter = "top", 
              options = list(scrollX = T))
```


## Mutation data (hg19)  
```{r}
query.maf.hg19 <- GDCquery(project = "TCGA-CHOL", 
                           data.category = "Simple nucleotide variation", 
                           data.type = "Simple somatic mutation", 
                           access = "open", 
                           legacy = T)

# check for maf availables  
DT::datatable(query.maf.hg19 %>% getResults() %>% dplyr::select(-contains("cases")), 
              filter = "top", 
              options = list(scrollX = T))
```


### Mutation data MC3 file  
This will download the MC3 MAF file from https://gdc.cancer.gov/about-data/publications/mc3-2017  
```{r}
maf <- getMC3MAF()
```


### Visualize the data 
To visualize the data you can use the Bioconductor package `maftools`  
```{r}
library(maftools)
library(tidyverse)

maf <- GDCquery_Maf(tumor = "CHOL", pipelines = "muse") %>% 
    read.maf()
maf

DT::datatable(getSampleSummary(maf), 
              filter = "top", 
              options = list(scrollX = T))
```


# Compilation of TCGA molecular subtypes  
TCGAbiolinks retrieved molecular subtypes information from TCGA samples. The functions `PanCancerAtlas_subtypes` and `TCGAquery_subtype` can be used to get the information tables.  

While the `PanCancerAtlas_subtypes` function gives access to a curated table retrieved from synapse (probably with the most updated molecular subtypes) the `TCGAquery_subtype` function has the complete table also with sample information from the TCGA marker paper.  

## `PanCancerAtlas_subtypes`: Curated molecular subtypes  
Data and description retrieved from synapse (https://www.synapse.org/#!Synapse:syn8402849)

Synapse has published a single file with all available molecular subtypes that have been described by TCGA (all tumor types and all molecular platforms), which can be accessed using the `PanCancerAtlas_subtypes` function as below:  

```{r}
subtypes <- PanCancerAtlas_subtypes()

DT::datatable(data = subtypes, 
              filter = "top", 
              options = list(scrollX = T))
```

The columns "Subtype_Selected" was selected as most prominent subtype classification (from the other columns).  

## `TCGAquery_subtype`: Working with molecular subtypes data  
The Cancer Genome Atlas (TCGA) Research Network has reported integrated genome-wide studies of various diseases. We have added some of the subtypes defined by these report in our package:  







