---
title: Choriocapillaris participate in the progress of age-related macular degeneration
  by regulating COL10A1 expressed by retina pigment epithelial
author: "Chunyou Yin^1\\*^; Chunhui Yin^2^"
csl: elsevier-with-titles-alphabetical.csl
output:
  bookdown::word_document2: default
  fig_caption: yes
  
link-citations: yes

bibliography:
- /home/yincy/git/R-codes/rmarkdown/library.bib
- /home/yincy/git/Data/mendeley.bib/library.bib
- /home/yincy/git/R-codes/rmarkdown/-mendeley.bib
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, 
                      messages = F, 
                      warning = F, 
                      fig.width = 8, 
                      fig.align = "center", 
                      dpi = 300)
```


```{r load-required-packages, message=FALSE, warning=FALSE, echo=FALSE}
library(tidyverse)
library(monocle3)
library(nichenetr)
library(cowplot)
```


## Abstract  
The well-known function of the choriocapillaris (CC) is supplying oxygen and nutrients to the out retina and retina pigment epithelial (RPE) while removing wastes. Besides oxygen and nutrients supplied by choriocapillaris is important for RPE's survival, signals that come from choriocapillaris should also important for maintaining retina pigment epithelial homeostasis. To date, most studies of the pathogenesis of age-related macular degeneration (AMD) are mainly focused on the RPE itself and largely ignored cells in RPE's micro-environment that might contribute to the progress of age-related macular degeneration. Recently Optical coherence tomography angiography (OCTA) and Histopathological studies found that changes have taken place at the choriocapillaris at the early stage of age-related macular degeneration. Understand how the changes take place at choriocapillaris and its impact on the RPE will contribute to the understanding of the pathogenesis of age-related macular degeneration. Here we take the first attempt to explore how choriocapillaris endothelial cell's change might contribute to the progress of neovascular age-related macular degeneration and find that TGFB1 expressed by choriocapillaris endothelial cells by influence the expression of COL10A1 expressed by retina pigment epithelial participate in the progress of neovascular age-related macular degeneration.  

> Key words: AMD; age-related macular degeneration; neovascular; choriocapillaris ; retina pigment epithelial; TGFB1; COL10A1  


## Introduction   
Age-related macular degeneration (AMD) is a leading cause of severe visual loss and blindness among the elderly population[@Wong2014a]. Its pathogenesis is poorly understood and it appears to involve a complex interaction of metabolic, functional, genetic, and environmental factors[@Evans2001;@Kaarniranta2019;@Kauppinen2016;@Wong2014a;@Fine1983;@Fritsche2013]. Late-stage AMD can be divided into two forms: geographic atrophy (GA; or dry) AMD, and neovascular (wet) AMD. The wet form of AMD comprises approximately 15% of the case, but because of its severity, accounts for the majority of severe visual loss due to AMD[@Gottlieb2020].   

Although anti-vascular endothelial growth factor (anti-VEGF) therapy has revolutionized the treatment of neovascular age-related macular degeneration. Some drawbacks have been reported, such as the treatment is not durable and visual gain was lost even after 2 years of treatment[@CATT_Research_Group2016], a significant proportion of patients developed into geographic atrophy after anti-VEGF therapy[@Grunwald2014;@Chakravarthy2013]. Genome-Wide Association Studies (GWAS) find dozens of sites associated with AMD, which implies complement activity, lipid metabolism, extracellular matrix remodeling, and angiogenesis is involved in the progress of AMD progression[@Fritsche2013;@Yu2011a], but the order which participates earlier in the progress of AMD is not determined and it is important for prevention and valuable for select therapeutic targets, but angiogenesis doesn't seem to be ahead of others.  

In light of the importance of RPE for the survival of photoreceptors, most of the researches about the pathogenesis of AMD is focused on the retina-pigment epithelial (RPE), and ignored the importance of the micro-environment in which RPE might be affected by nearby cells. Histopathological and the recent advancement in imaging technology, Optical coherence tomography (OCT), have found that in the very early stage of AMD, notable changes have been identified at the choriocapillaris, such as reduced blood flow, choriocapillaris dropout[@Sohn2019;@Chatziralli2018;@Seddon2016;@Parviz2017;@Wakatsuki2015]. This implies that the changes that take place at CC are ahead of RPE degeneration and it might responsible for the onset of AMD. But the mechanism of how CC changes might affect retina degeneration is unexplored to our knowledge. In light of the above findings, here we trying to understand how CC might affect RPE in the progress of AMD, and find that TGFB1 expressed by CC endothelial cells by affecting the expression of AMD related gene, COL10A1, in RPE and this might contribute to the progress of neovascular AMD.  


## Methods  
### Statistical and Data Analysis  
To find data suitable for our purpose, we searched the NCBI Gene Expression Omnibus database (GEO), and find the CD31 enriched single-cell RNA-seq of choroidal described by Voigt[@Voigt2019a], which characterized the arteriole, vein, and choriocapillaris endothelial cells, and the choriocapillaris endothelial cells are used to define the potential ligands expressed by the choriocapillaris endothelial cells that might regulator macular degeneration related genes expressed by RPE. The single-cell trajectory was conducted on choriocapillaris endothelial cells among CD31 enriched patients to find the differentially expressed ligands along the trajectory using monocle3[@Trapnell2014a;@Qiu2017;@Qiu2017a], batch effects were removed using the method described by Haghverdi[@Haghverdi2018].  

For the RPE data, we use the bulk RNA-seq data mentioned by Kim[@Kim2018a], which they find a significant impact on the anti-sense transcripts between the normal and early AMD patients compare to the sense transcripts. This data is used to define the background expressed genes and age-related macular degeneration genes in early AMD. Differential expression analysis was carried out with edgeR-limma[@McCarthy2012;@Ritchie2015] R/Bioconductor packages between normal and early AMD patients.  

### Compliance with Ethics Guidelines   
All procedures performed in this studies were in accordance with the ethical standards of the of the institusion and/or national research committee and with the Declaration of Helsinki, as revised in 2013. 

### Data and Code availability  
CD31 enriched choriocapillaris endothelial single-cell RNA-seq data: GSE139522.  
bulk RNA-seq health and early-stage AMD data: GSE99248.  
code and data is available at GitHub: https://github.com/YinCY123/AMD  


## Results  
### ligand-target regulatory pairs  
Before inferring ligand-target regulatory pairs, ligands activity needs to be performed on the age-related macular degeneration genes, in other words, assess how well each CC endothelial cells expressed ligands can predict the RPE expressed age-related macular degeneration related genes compared to the background of expressed genes. The top 15 highly active ligands are used to infer ligand-target regulation potential pairs (Figure \@ref(fig:ligand-target-pairs)A).  

Based on prior knowledge on signaling and gene regulatory networks, by using the network propagation method[@Browaeys2019a] on the integrated networks to infer active ligands expressed by CC endothelial cells and their gene regulatory effects on RPE, we find few high potential ligand-target regulatory pairs, based on the top 15 highly active ligands (Figure \@ref(fig:ligand-target-pairs)B).  

```{r load-data}
ligand_target_heatmap <- readRDS(file = "../Retina/res/ligand_target_heatmap.rds")
ligand_activity_heatmap <- readRDS(file = "../Retina/res/ligand_activity_heatmap.rds")
```


```{r ligand-target-pairs-no-eval, message=FALSE, warning=FALSE, fig.cap="(A) The activity of top 15 ligands. (B) Ligand-target paris identified by the network propagation methods.", eval=FALSE}
plot_grid(
  ligand_activity_heatmap +
    ggtitle(label = "ligand activity") +
    theme(plot.title = element_text(size = 9, face = "bold", hjust = "0.5"), 
          legend.position = "right"),
  
  ligand_target_heatmap +
  theme(legend.position = "top", 
        plot.title = element_text(size = 12, face = "bold", hjust = 0.5)), 
  nrow = 1, 
  rel_widths = c(1.8, 5.0), 
  labels = c("A", "B"))
```

### TGFB1 upregulated in choriocapillaris endothelial cells among early AMD patients  
To find out which ligands are differentially expressed between normal and early AMD patients, we use the CD31 enriched choriocapillaris endothelial cells data[@Voigt2019a] and find that TGFB1 expression higher in the early stage neovascular AD patients compared to the health (Figure \@ref(fig:expression-of-ligands)A). Then single-cell trajectory analysis was conducted according to the standard precedure[@Trapnell2014a;@Qiu2017;@Qiu2017a], and we indeed find TGFB1 differentially expressed along the pseudotime trajectory, which means from the normal state to a more stress state (early AMD, endothelial cells dropout) TGFB1's expression is becoming higher (Figure \@ref(fig:expression-of-ligands)B).
```{r load-single-cell-image, message=FALSE, warning=FALSE}
cc <- readRDS(file = "/home/yincy/git/R-codes/Retina/res/cc.rds")
expression_ligands_ridges <- readRDS(file = "../Retina/res/expression_ligands_ridges.rds") +
  scale_y_discrete(name = "top ligands")
```

```{r single-cell-plot, message=FALSE, warning=FALSE}
pseudotime_p <- plot_cells(cds = cc, 
           color_cells_by = "pseudotime", 
           show_trajectory_graph = T, 
           cell_size = 1.2, 
           trajectory_graph_segment_size = 1, 
           label_leaves = F, 
           label_roots = F)

percent.mt_p <- plot_cells(cds = cc, 
                           color_cells_by = "percent.mt", 
                           cell_size = 1.2, 
                           trajectory_graph_segment_size = 1, 
                           show_trajectory_graph = T, 
                           label_roots = F, 
                           label_leaves = F)

expression_of_TGFB1 <- plot_cells(cds = cc, 
                                  reduction_method = "UMAP", 
                                  genes = "TGFB1", 
                                  show_trajectory_graph = T, 
                                  trajectory_graph_segment_size = 0.2, 
                                  cell_size = 1, 
                                  min_expr = 0, 
                                  label_leaves = F, 
                                  label_roots = F) +
  scale_color_gradient(name = "relative\nexpression", low = "green", high = "red")

expression_of_COL18A1 <- plot_cells(cds = cc, 
                                  reduction_method = "UMAP", 
                                  genes = "COL18A1", 
                                  show_trajectory_graph = T, 
                                  trajectory_graph_segment_size = 0.2, 
                                  cell_size = 1, 
                                  min_expr = 0, 
                                  label_leaves = F, 
                                  label_roots = F) +
  scale_color_gradient(name = "relative\nexpression", low = "green", high = "red")
```

```{r expression-of-ligands-no-eval,fig.cap = "(A) The expression of ligands between early AMD patients (n = 3) and health patients (n = 7). The more shift to the right the higher of the expression. (B) The expression of TGFB1 along the trajectory, from the top left to the bottom right the expression of TGFB1 is becoming more higher, which is from normal state to early AMD state.", message=FALSE, warning=FALSE, eval=FALSE}
plot_grid(expression_ligands_ridges, 
          expression_of_TGFB1, 
          labels = c("A", "B"))
```

To confirm our result is solid, we calculate and plot the percentage of mitochondrial genes in each cell (Figure \@ref(fig:pseudotime-and-percent-mt)B), indeed we find an elevated percentage of mitochondrial genes along the trajectory, this is consistent with previous reports that choriocapillaris is thinning and dropout in the early stage of AMD[@Chirco2017;@Chatziralli2018;@Lengyel2004;@Sohn2014a] and the percentage of mitochondrial genes is elevated in more stress or dying cells[@Seurat2019;@Islam2014;@Ilicic2016].  

```{r pseudotime-and-percent-mt-no-eval, fig.cap = "(A) The status of the choriocapillaris endothelial cells changes along the pseudotime trajectory. (B) The percentage of mitochandrial genes expressed by the choriocapillaris endothelial cells along the pseudotime trajectory, which is becoming higher in the early AMD state.", eval=FALSE}
plot_grid(pseudotime_p, 
          percent.mt_p,
          labels = c("A", "B"))
```


### COL10A1 differentially expressed among health and early AMD patients  
Differential expression analysis between health and early AMD patients was conducted following the standard pipeline of edgeR-limma[@McCarthy2012;@Ritchie2015] R/Bioconductor package, and we find that TLR4 and COL10A1 were significantly differential expressed between early AMD and health patients (Figure \@ref(fig:expression-of-target)). TLR4 participates in the activation of the innate immune system which is consistent with previous Genome-Wide Association studies about the participation of the immune system in the pathogenesis of AMD[@Fritsche2013;@Liao2019;@Whitmore2015], although we didn't notice the difference of IL1B along the trajectory, it might be the immune cells or pericyte in the choroidal secreted the IL1B that affect the expression of TLR4.  

According to the classification of Hogan[@Hogan1961], Bruch's membrane consists of five layers: the basement membrane of the RPE, the inner collagenous layer (ICL), the elastin layer (EL), the outer collagenous layer (OCL) and, the basement membrane of the choriocapillaris. Although it is not exactly clear how Bruch's membrane is formed, gene expression data indicated that both the choroid and RPE cells are capable of synthesizing the major components of Bruch's membrane[@Booij2009;@Voigt2019a;@Orozco2020;@Kim2018a]. As Bruch's membrane is a dynamic tissue[@Booij2010a], the lowered expression of COL10A1 could change the structure of Bruch's membrane and that might responsible for the deposition of drusen in the Bruch's membrane. Besides affecting the structure of Bruch's membrane, several experiments have reported the proteolytic fragments cleaved from collages exhibit anti-angiogenic and anti-tumor properties[@Su2004;@Oreilly1997], and downregulation of COL10A1 may disturb the balance between inhibitor and simulator of angiogenesis signal and cause neovascularization, which is the character of wet AMD.  

To find which transcription regulator might responsible for the downregulate of COL10A1, we integrated the TGFB1 and COL10A1 signaling networks (Figure \@ref(fig:TGFB1-COL10A1-signaling)) and find three transcription factors RUNX2, AR, and ESR1 that could regulate the expression of COL10A1. Next, we examined the expression of these regulators and find that only RUNX2 and ESR1 expressed after filtering, and RUNX2 was significantly down-regulated between health and early AMD patients (Figure \@ref(fig:expression-regulators)), this is consistent with the find of Li[@Li2011], which found that RUNX2 can promote the expression of COL10A1 in mice, and the lowered expression of RUNX2 could result in the lowered expression of COL10A1.  

```{r expression-of-target-no-eval, message=FALSE, warning=FALSE, fig.cap="The expression of target genes in early AMD. TLR4, COL10A1 were significantly differental expressed between normal(n = 3) and early AMD patients(n = 7) but VEGF, TIMP3 and TGFBR1 were not.", eval=FALSE}
expression_targets_RPE_boxplot <- readRDS(file = "../Retina/res/expression_targets_RPE_boxplot.rds")
expression_targets_RPE_boxplot
```


## Discussion  
Although anti-VEGF has brought light for neovascular AMD patients, some side-effects have been reported[@CATT_Research_Group2016;@Grunwald2014;@Chakravarthy2013]. According to the Genome-Wide Association Studies[@Fritsche2013;@Yu2011a], complement activity, lipid metabolic, extracellular matrix remodeling, and angiogenesis are involved in the pathogenesis of AMD, but which factor onset ahead of others is not determined, and this is important for prevention and choose the optimal candidate of therapeutic targets. However angiogenesis doesn't seem to be ahead of others, this implies inhibit neovascularization may not be the best choice for neovascular AMD. Here we find that COL10A1, a member of the collagen family, which participates in extracellular matrix remodeling, might a better therapeutic target than anti-VEGF for neovascular AMD. Besides participates in extracellular remodeling, which might responsible for the deposition of drusen, many studies have found a number of proteolytic fragments cleaved from collagen have anti-angiogenesis and anti-tumor properties[@Oreilly1997]. From this point of view, neovascularization might be the imbalance of stimulating and inhibit signals of angiogenesis. In light of these findings, we anticipate that the changed expression of COL10A1 in one aspect could change the structure of Bruch's membrane, and cause the deposition of drusen in the Bruch's membrane, the other way it might imbalanced stimulate and inhibit signals of angiogenesis and that might responsible of the neovascularization.  

TLR4, an AMD related gene, also significantly differentially expressed, which participate in activating the innate immune system, but we didn't see the differentially expressed of its ligand IL1B in choriocapillaris endothelial cells, perhaps it produced by other cells in the choroidal like monocytes, neutrophils infiltrate to the inflammation site. As it is significantly differentially expressed at the early stage of AMD, this implies the innate immune system is also very important for the pathogenesis of AMD and more researches need to be done about how the immune system participates in the progression of AMD.  

Another collagen family member, COL18A1, is also deferentially expressed along the trajectory, which has been considered as a critical factor for retina structure[@Sertie2000]. Besides affecting the retina structure studies also reveal the anti-tumor and anti-angiogenesis property of COL18A1[@Oreilly1997], which on the other side implies the imbalanced angiogenesis signal, and extracellular matrix remodeling participate in the progress of AMD.  

Another important issue to address is the order of complement activity, lipid metabolic, extracellular matrix remodeling, and angiogenesis which has been identified by Genome-Width Association Studies that participate in the pathogenesis of AMD, this is critical for choosing the optimal therapeutic targets. With the advance of Next-generation Sequencing (NGS) and Metabolomics/proteomics, more research needs to be done to determine the order.  


## Author information  
### Affiliations  
1. School of Medicine, Zhejiang University, Hangzhou, Zhejiang 310009, China    
2. 51 Job, Shanghai Pudong New Area 201210, China  

### Author Contribution  
Chunyou conceive, design, conducted the data analysis, and write the manuscripts, Chunhui helps build the computational environment and some help with dimension reduction algorithm.   

### Corresponding authors  
Correspondence to Chunyou Yin (yinchunyou@zju.edu.cn)

## Conflict of interests  
The authors declare no competing interests.  

## Grant information  
The authors receive no funds.  


## Figures and Tables  
```{r ligand-target-pairs, message=FALSE, warning=FALSE, fig.cap="(A) The activity of top 15 ligands. (B) Ligand-target paris was identified by the network propagation methods."}
plot_grid(
  ligand_activity_heatmap +
    ggtitle(label = "ligand activity") +
    theme(plot.title = element_text(size = 9, face = "bold", hjust = "0.5"), 
          legend.position = "right"),
  
  ligand_target_heatmap +
  theme(legend.position = "top", 
        plot.title = element_text(size = 12, face = "bold", hjust = 0.5)), 
  nrow = 1, 
  rel_widths = c(1.8, 5.0), 
  labels = c("A", "B"))
```

```{r expression-of-ligands,fig.cap = "(A) The expression of ligands between early AMD patients (n = 3) and health patients (n = 7). The more shift to the right the higher the expression. (B) The expression of TGFB1 along the trajectory, from the top left to the bottom right the expression of TGFB1 is becoming higher, which is from normal state to early AMD state.", message=FALSE, warning=FALSE}
plot_grid(expression_ligands_ridges, 
          expression_of_TGFB1, 
          labels = c("A", "B"))
```

```{r pseudotime-and-percent-mt, fig.cap = "(A) The status of the choriocapillaris endothelial cells changes along the pseudotime trajectory. (B) The percentage of mitochondrial genes expressed by the choriocapillaris endothelial cells along the pseudotime trajectory, which is becoming higher in the early AMD state."}
plot_grid(pseudotime_p, 
          percent.mt_p,
          labels = c("A", "B"))
```

```{r expression-of-target, message=FALSE, warning=FALSE, fig.cap="The expression of target genes in early AMD. TLR4, COL10A1 were significantly differential expressed between normal(n = 3) and early AMD patients(n = 7) but VEGF, TIMP3 and TGFBR1 were not."}
expression_targets_RPE_boxplot <- readRDS(file = "../Retina/res/expression_targets_RPE_boxplot.rds")
expression_targets_RPE_boxplot
```

## Supplementary Materials  
### Statistic of ligands expression along the trajectory  
```{r statistic-result-ligands, fig.cap="The statistical significance of ligands expressed along the trajectory in the choriocapillaris endothelial cells."}
cc_graph_test_res <- read.csv(file = "/home/yincy/git/R-codes/Retina/res/cc_graph_test_res.csv")

top_ligands <- c("IL1B", "TGFB1", "ADAM17", "IL33", 
                 "CXCL12", "CXCL2", "COL18A1", "CXCL3", 
                 "IHH", "OCLN", "HMGB1", "FGF2", 
                 "APP", "HSP90B1", "BMP2")

o <- cc_graph_test_res %>% 
  dplyr::filter(gene_short_name %in% top_ligands) %>% 
  arrange(q_value) %>% 
  pull(gene_short_name)

cols <- rep("grey30", length(o))
names(cols) <- o
cols["TGFB1"] <- "blue"

cc_graph_test_res %>% 
  dplyr::filter(gene_short_name %in% top_ligands) %>% 
  ggplot(aes(reorder(gene_short_name, q_value), -log10(q_value))) +
  geom_bar(stat = "identity", fill = cols) +
  scale_y_continuous(name = "-log10(q_value)", breaks = seq(0, 50, 5)) +
  scale_x_discrete(name = NULL) +
  coord_flip() +
  theme_classic() +
  ggtitle(label = "significance of ligands along trajectory")
```


### Statistic of targets expression  
```{r statistic-result-targets, message=FALSE, warning=FALSE, fig.cap="The statistical significance of AMD related genes between health and AMD patients."}
statistic_significance_targets <- readRDS(file = "/home/yincy/git/R-codes/Retina/res/statistic_significance_targets.rds")

plot_grid(statistic_significance_targets + theme_classic(), 
          nrow = 2, 
          rel_heights = rep(1, 2)) 
```


### signal network of TGFB1 COL10A1  
```{r TGFB1-COL10A1-signaling, fig.align="center", fig.cap="The signaling pathway of TGFB1-COL10A1 pairs, which implies RUNX2, ESR1 and AR are the potential transcription factors that could regulate the expression of COL10A1."}
knitr::include_graphics(path = "/home/yincy/git/R-codes/Retina/res/TGFB1-COL10A1-signaling-network.png")
```


### expression of COL18A1  
```{r, fig.cap="The expression of COL18A1."}
plot_grid(expression_of_COL18A1, 
          ncol = 2)
```


### expression of regulators  
```{r expression-regulators, fig.cap="The expression of transcription regulators between health and AMD patients."}
expression_regulators <- readRDS(file = "/home/yincy/git/R-codes/Retina/res/expression_regulators.rds")
plot_grid(expression_regulators, 
          ncol = 2)
```


### Statistic of regulator expression  
```{r statistic-targets, message=FALSE, warning=FALSE, fig.cap="The statistic expression of regulators that could regulate the expression of COL10A1, after filtering only RUNX2, and ESR1 are expressed among health and AMD patients, only RUNX2 significantly expressed between health(n = 3) and AMD patients(n = 7)."}
statistic_significance_regulator <- readRDS(file = "/home/yincy/git/R-codes/Retina/res/statistic_significance_regulator.rds")
plot_grid(statistic_significance_regulator, 
          nrow = 2)
```


### dimensional reduction of bulk-seq health and early AMD patients  
```{r, fig.cap="Dimensional reduction of bulk-seq normal and early AMD patients, which show the high quality of the sequencing data, as normal and early AMD patients can be clearly separated."}
knitr::include_graphics(path = "../Retina/res/dimreduc_bulk_normal_amd.png")
```


### Software and R/Bioconductor packages  
Name               | website 
-------------------|---------------------------------------------------------------
R                  | https://www.r-project.org
RStudio            | https://rstudio.com/
edgeR              | http://www.bioconductor.org/packages/release/bioc/html/edgeR.html
limma              | http://www.bioconductor.org/packages/release/bioc/html/limma.html
monocle3           | http://cole-trapnell-lab.github.io/monocle-release/
tidyverse          | https://www.tidyverse.org/
cowplot            | https://github.com/wilkelab/cowplot
RColorBrewer       | https://CRAN.R-project.org/package=RColorBrewer
nichenetr          | https://github.com/saeyslab/nichenetr
GEOquery           | http://www.bioconductor.org/packages/release/bioc/html/GEOquery.html
ggridges           | https://CRAN.R-project.org/package=ggridges
EnsDb.Hsapiens.v86 | http://www.bioconductor.org/packages/release/data/annotation/html/EnsDb.Hsapiens.v86.html
knitr              | https://yihui.org/knitr/
rmarkdown          | https://github.com/rstudio/rmarkdown
bookdown           | https://github.com/rstudio/bookdown


### sessionInfo
```{r sessionInfo}
sessionInfo <- readRDS(file = "/home/yincy/git/R-codes/Retina/res/sessionInfo.rds")
sessionInfo
```


## References  
