---
title: "journal"
author: 
- Chunyou Yin
date: 2020-05-29
output: 
    html_document:
        toc: false
        highlight: tango
        
bibliography: [library.bib;-mendeley.bib]  

csl: optometry-and-vision-science.csl
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, messages = F, warning = F)
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(monocle3)
library(nichenetr)
library(cowplot)
```


### Abstract  
The well known function of chroiocapillaris (CC) is the supplying oxygen and nutrients to the out retina and retina pigment epithelial (RPE) while remove waste. Besides oxygen and nutrients supplied by CC is important for RPE's survive, signals come from CC should also important for maintain RPE's homeostatsis. To date, most studies of the pathgenesis of age-related macular degeneration (AMD) are mainly focused on the RPE itself and largely ignored cells in RPE's microenvironment that might contribute to the progress of AMD. Recently Optical coherence tomography angiography (OCTA) and Histopathological studies found that changes have taken place in the CC at the early stage of AMD. Understand how the changes take place at CC and it's impact on the RPE will contribute to the understand of pathgenesis of age-related macular degeneration. Here we take the first attempt to explor how CC endothelial cells's change might contribute to the progress of AMD and find that TGFB1 expressed by choriocapillaris endothelail cells by regulate COL10A1 gene participate in the pregression of AMD.  

> Key words: age-related macular degeneration; choriocapillaris ; retina pigment epithelial 


### Introduction   
Age-related macular degeneration (AMD) is an leading cause of sever visual loss and blindness among the eldly population[@Wong2014a]. It's pathgenesis is poorly understood and it apears to involve a complex interaction of metabolic, functional, genetic and environment factors[@Evans2001;@Kaarniranta2019;@Kauppinen2016;@Wong2014a;@Fine1983;@Fritsche2013]. Genome-wide association studies (GWAS) conducted by The AMD Gene Consortium find numerous sites associated with AMD, which implies complement activity, lipid metabolism, extracellular matrix remodeling and angiogenesis is involved in the progress of AMD progression[@Fritsche2013].  

Late-stage AMD can be divided into two forms: geographic atrophy (GA; or dry) AMD, and neovascular (wet) AMD. The wet form of AMD comprises approcimately 15% of the case, but because of its severity, accounts for the majority of servere visual loss due to AMD[@Gottlieb2020]. Although anti-vascular endothelial growth factor (anti-VEGF) therapy has revolutionized the treatment of neovascular age-related macular degeneration. Some drawbacks have been reported, such as the treatment is not durable and visual gain was lost after 2 years treatment[@CATT_Research_Group2016], significant proportion of patients developed into geograpic atrophy after anti-VEGF therapy[@Grunwald2014;@Chakravarthy2013]. 

In light of the importance of RPE for the survive of photoreceptors, most of the research about the pathgenesis of AMD are focused on the retina-pigment epithelial (RPE), and ignored the importance of microenvirnment in which RPE might be affect by other cells. Histopathological and the recent advancement in imaging technology, Optical coherence tomography (OCT), have found that in the very early stage of AMD, notable changes have identified at the choriocapillaris, such as reduced blood flow, choriocapillaris dropout[@Sohn2019;@Chatziralli2018;@Seddon2016;@Parviz2017;@Wakatsuki2015]. This implies that the changes take place at CC is ahead of RPE and it might responsible for the onset of AMD. But the mechinism of how CC changes may affact retina degeneration is largely unexplored. In the light of above findings, here we trying to understand how CC might affact RPE in the progress of AMD, and find that TGFB1 expressed by CC endothelial cells affact the expression of AMD related gene COL10A1 in RPE and this might contribute to the progression of AMD.  


### Results  
#### Data and Statictic analysis  
To find data suitable for our purpose, we searched the Gene Expression Omnibus database (GEO), and find the CD31 enriched single-cell RNA-seq of choriodal described by Voigt[@Voigt2019a], this data is used to define the potential ligands expressed by the choriocapillaris endothelial cells that might regulator macular degeneration related genes expressed by RPE. Single cell trajectory were conducted on choriocaplaris endothelial cells among CD31 enriched macular samples to find the differentially expressed ligands along the trajectory using monocle3[@Trapnell2014;@Qiu2017;@Qiu2017a], batch effects is removed using method described by Haghverdi[@Haghverdi2018].  

For the RPE data, we use the bulk RNA-seq data mentioned by Kim[@Kim2018a], which they describe the siginificant impact on the anti-sense transcripte between the normal and AMD patients. This data is used to define the background expressed genes and expressed age-related macular degeneration genes at early AMD. Differential expression analysis was carried out with edgeR-limma[@McCarthy2012;@Ritchie2015] R/Bioconductor packages between normal and early AMD samples.  


#### ligand target paris  
Before to infer ligand-target links, ligands activity needs to be performed on the age-related macular degeneration genes, in other words assess how well each CC endothelial cell expressed ligand can predict the RPE expressed age-related macular degeneration related genes compared to the background of expressed genes. The top 15 highly active ligands are show in fig1.  

Based on prior knowlege on signaling and gene regulatory networks, by using network propagation methods on the integrated networks to infer active ligands expressed by CC endothelial cells and their gene regulatory effects on RPE using methods metioned by Browaeys[@Browaeys2019], we find few high potential ligand-target  regulatory paris, based on the top 15 highly active ligands and find that TGFB1 and COL10A1 among these high regulatory potential pairs.  
```{r, echo=FALSE, message=FALSE}
ligand_target_heatmap <- readRDS(file = "../Retina/res/ligand_target_heatmap.rds")
ligand_activity_heatmap <- readRDS(file = "../Retina/res/ligand_activity_heatmap.rds")

plot_grid(
  ligand_activity_heatmap +
    ggtitle(label = "ligand activity") +
    theme(plot.title = element_text(size = 9, face = "bold", hjust = "0.5"), 
          legend.position = "right", 
          legend.box.spacing = unit(0, "cm")),
  ligand_target_heatmap +
  theme(legend.position = "top", 
        plot.title = element_text(size = 12, face = "bold", hjust = 0.5)), 
  nrow = 1, 
  rel_widths = c(1.8, 5))
```


#### COL10A1 differentially expressed among normal and early AMD patients  
Differential expression analysis between normal and early AMD patients was conducted following the standard pepline of edgeR-limma[@McCarthy2012;@Ritchie2015] R/Bioconductor package. TLR4 and COL10A1 were siginificantly differential expressed among early AMD and normal patients. TLR4 is participate in the activation of innate immune system which is consistent with previous study about the participate of immune system in the pathgenesis of AMD[@Fritsche2013;@Liao2019;@Whitmore2015].  

COL10A1 has been identified associated with age-related degeneration[@Yu2011a;@Fritsche2013], 
Bruch's membrane can be classified into five layers: the basement membrane of the RPE, the inner collagenous layer (ICL), the elasin layer (EL), the outer collagenous layer (OCL) and, the basement membrane of the choriocapillaris[@Hogan1961]. As Bruch's membrane is an highly dynamic acellular tissue[@Booij2010a], down regulation of COL10A1 might change the structure of Bruch's membrane, which will affect the oxygen and nutrient flow through it to RPE and contribute to the degeneration of RPE, and possibly responsible for the deposition of drusen at Bruch's membrane[@Crabb2002;@]. 
```{r, echo=FALSE, message=FALSE, warning=FALSE}
expression_targets_RPE_boxplot <- readRDS(file = "../Retina/res/expression_targets_RPE_boxplot.rds")
expression_targets_RPE_boxplot
```


#### TGFB1 differentially expressed along trajectory  

```{r, message=FALSE, warning=FALSE, echo=FALSE}
load(file = "../../Data/AMD-related/AMD-single-cell/monocle3_amd_single_cell.RData")
expression_ligands_ridges <- readRDS(file = "../Retina/res/expression_ligands_ridges.rds")
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
pseudotime_p <- plot_cells(cds = cc, 
           color_cells_by = "pseudotime", 
           show_trajectory_graph = T, 
           cell_size = 1.2, 
           trajectory_graph_segment_size = 0.1)

percent.mt_p <- plot_cells(cds = cc, 
                           color_cells_by = "percent.mt", 
                           cell_size = 1.2, 
                           trajectory_graph_segment_size = 0.1, 
                           show_trajectory_graph = T)

expression_of_TGFB1 <- plot_cells(cds = cc, 
                                  reduction_method = "UMAP", 
                                  genes = "TGFB1", 
                                  show_trajectory_graph = T, 
                                  trajectory_graph_segment_size = 0.1, 
                                  cell_size = 1.2, 
                                  min_expr = 0) +
  scale_color_gradient(name = "relative\nexpression", low = "green", high = "red")

plot_grid(pseudotime_p, 
          percent.mt_p)
```


```{r, echo=FALSE, message=FALSE, warning=F}
plot_grid(expression_ligands_ridges, 
          expression_of_TGFB1)
```


### Discussion  


### Data and code  
CD31 enriched choriocapillaris endothelial single-cell RNA-seq data: GSE139522.  
bulk RNA-seq RPE data: GSE99248 
code is available at GitHub: https://github.com/YinCY123/AMD


### Acknowledgements   




### References  


### Supplimentary Materials  

