################################################
#
#  Learning pathview for pathway visulisation
#  YinCY
#  2018/09/04
#  yinchunyou@zju.edu.cn
#
################################################

library(pathview)

########################################
#  Quick Start with demo data
########################################

pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id = '04110',
                   species = 'hsa', out.suffix = 'gse16873')

########################################
# Common uses for data visualization
########################################

data("gse16873.d")
filename <- system.file('extdata/gse16873.demo', package = 'pathview')
gse16873 <- read.delim(file = filename, row.names = 1)

gse16873.d <- gse16873[, 2*(1:6)] - gse16873[, 2*(1:6) - 1]
head(gse16873.d)

data("demo.paths")
data("paths.hsa")
head(paths.hsa, n = 5)


i <- 1
pv.out <- pathview(gene.data = gse16873.d[, 1], pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa', out.suffix = 'gse16873', kegg.native = TRUE)

list.files(path = getwd(), pattern = 'hsa04110')

str(pv.out)
head(pv.out$plot.data.gene)

# Graph from the first example above has a single layer. Node colors were 
# modified on the original graph layer, and original KEGG node labels were
# kept intact. This way the output file size is as small as the original
# KEGG PNG file, but the computing time is relative long. 
# If we want a fast view and do not mind doubling the output file size, we
# may do a two-layer graph with same.layer = F. This way node colors and labels
# are added on an extra layer above the original KEGG graph.

# Note that the original KEGG gene labels (or EC numbers) were replaced by 
# official gene symbols.

pv.out <- pathview(gene.data = gse16873.d[, 1], 
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa', 
                   out.suffix = 'gse16873.2layer', 
                   kegg.native = TRUE,
                   same.layer = F)

pv.out <- pathview(gene.data = gse16873.d[, 1], 
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa',
                   out.suffix = 'gse16873',
                   kegg.native = FALSE,
                   sign.pos = demo.paths$sel.paths[i])

dim(pv.out$plot.data.gene)
head(pv.out$plot.data.gene)


# put graph and legend in two pages
pv.out <- pathview(gene.data = gse16873.d[, 1],
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa',
                   out.suffix = 'gse16873.2layer',
                   kegg.native = FALSE,
                   sign.pos = demo.paths$spos[i],
                   same.layer = FALSE)


# In Graphviz view, we have more control over the graph layout. We may
# split the node groups into individual detached nodes. 
# We may even expand the multiple-gene nodes into individual genes.

pv.out <- pathview(gene.data = gse16873.d[, 1], 
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa', 
                   out.suffix = 'gse16873.split',
                   kegg.native = FALSE,
                   sign.pos = demo.paths$sel.paths[i],
                   split.group = TRUE)
dim(pv.out$plot.data.gene)
head(pv.out$plot.data.gene)


pv.out <- pathview(gene.data = gse16873.d[, 1],
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa',
                   out.suffix = 'gse16873.split.expand',
                   kegg.native = FALSE,
                   sign.pos = demo.paths$sel.paths[i],
                   split.group = TRUE,
                   expand.node = TRUE)

dim(pv.out$plot.data.gene)
head(pv.out$plot.data.gene)


###############################
#  Data integration
###############################

# pathview provides strong support for data integration. It can be used to 
# integrate, analyze and visulize a wide variety of biological data: gene
# expression, protein expression, genetic association, metabolite, genomic 
# data literature, and other data types mappable to pathways.

# Noteably, it can be directly used for metagenomic, microbiome or unknown
# species data when the data are mapped to KEGG ortholog pathway.


sim.cpd.data <- sim.mol.data(mol.type = 'cpd', nmol = 3000)
head(sim.cpd.data)
data("cpd.simtypes")

# We generate a native KEGG view graph with both gene data and compound data.
# Such metabolic pathway graphs generated by pathview is the same as the 
# original KEGG graphs, except that the compound nodes are magnified for better
# view of the colors.

i <- 3
print(demo.paths$sel.paths[i])

pv.out <- pathview(gene.data = gse16873.d[, 1],
                   cpd.data = sim.cpd.data,
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa',
                   out.suffix = 'gse16873.cpd',
                   keys.align = 'y',
                   kegg.native = TRUE,
                   key.pos = demo.paths$sel.paths[i])
str(pv.out)
head(pv.out$plot.data.cpd)

pv.out <- pathview(gene.data = gse16873.d[, 1], 
                   cpd.data = sim.cpd.data,
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa',
                   out.suffix = 'gse16873.cpd',
                   keys.align = 'y',
                   kegg.native = FALSE,
                   key.pos = demo.paths$kpos2[i],
                   sign.pos = demo.paths$spos[i],
                   cpd.lab.offset = demo.paths$offs[i])

###############################
# Multiple states or samples
###############################

set.seed(10)
sim.cpd.data2 <- matrix(sample(x = sim.cpd.data, size = 18000, replace = TRUE), ncol = 6)

rownames(sim.cpd.data2) <- names(sim.cpd.data)
colnames(sim.cpd.data2) <- paste('exp', 1:6, sep = '')
head(sim.cpd.data2)


pv.out <- pathview(gene.data = gse16873.d[, 1:3],
                   cpd.data = sim.cpd.data2[, 1:2],
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa',
                   out.suffix = 'gse16873.cpd.3-2s',
                   keys.align = 'y',
                   kegg.native = TRUE,
                   match.data = FALSE,
                   multi.state = TRUE, 
                   same.layer = TRUE)

head(pv.out$plot.data.cpd)


# KEGG view with data match
pv.out <- pathview(gene.data = gse16873.d[, 1:3],
                   cpd.data = sim.cpd.data2[, 1:2],
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa',
                   out.suffix = 'gse16873.cpd.3-2s.match',
                   keys.align = 'y',
                   kegg.native = TRUE,
                   match.data = TRUE,
                   multi.state = TRUE,
                   samme.layer = TRUE)
head(pv.out$plot.data.cpd)


# graphviz view
pv.out <- pathview(gene.data = gse16873.d[, 1:3],
                   cpd.data = sim.cpd.data2[, 1:2],
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa',
                   out.suffix = 'gse16873.cpd.3-2s',
                   keys.align = 'y',
                   kegg.native = FALSE,
                   match.data = FALSE,
                   multi.state = TRUE,
                   same.layer = TRUE,
                   key.pos = demo.paths$kpos2[i],
                   sign.pos = demo.paths$spos[i])

# plot the samples separately, one sample per graph

# Note that in this case, the samples in gene.data and cpd.data has to be 
# matched as to assigned to the same graph. Hence, argument match.data isn't
# really effective here.

## Plot samples/states separately
pv.out <- pathview(gene.data = gse16873.d[, 1:3], 
                   cpd.data = sim.cpd.data2[, 1:2], 
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa', 
                   out.suffix = 'gse16873.cpd.3-2s',
                   keys.align = 'y',
                   kegg.native = TRUE,
                   match.data = FALSE,
                   same.layer = TRUE)

# as described above, KEGG views on the same layer may takes some time.
# Again, we can choose to do KEGG view with two layers as to speed up the
# process iif we don't mind losing the original KEGG gene labels (or EC numbers).
pv.out <- pathview(gene.data = gse16873.d[, 1:3],
                   cpd.data = sim.cpd.data2[1:2],
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa',
                   out.suffix = 'gse16873.cpd.3-2s.2layer',
                   keys.align = 'y',
                   kegg.native = TRUE,
                   match.data = F,
                   multi.state = T,
                   same.layer = F)

################################
#  Discrete data
################################

# So far, we have been dealing with continuous data. But we often work with 
# discrete data too. For instance, we select list of signficant genes or 
# compound based on some statistics(p-value, fold change etc). The input data 
# can be named vector of two levels, either 1 or 0 (signficant or not), or it 
# can be a shorter list of signficant gene/compound names. 

library(org.Hs.eg.db)
gse16873.t <- apply(X = gse16873.d, MARGIN = 1, FUN = function(x) t.test(x, alternative = 'two.sided')$p.value)

sel.genes <- names(gse16873.t)[gse16873.t < 0.01]
sel.cpds <- names(sim.cpd.data)[abs(sim.cpd.data) > 0.5]
pv.out <- pathview(gene.data = sel.genes,
                   cpd.data = sel.cpds,
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa',
                   out.suffix = 'sel.genes.sel.cpd',
                   keys.align = 'y',
                   kegg.native = TRUE,
                   key.pos = demo.paths$kpos1[i],
                   limit = list(gene = 5, cpd = 2),
                   bins = list(gene = 5, cpd = 2),
                   na.col = 'gray',
                   discrete = list(gene = TRUE, cpd = TRUE))


pv.out <- pathview(gene.data = sel.genes,
                   cpd.data = sim.cpd.data,
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa',
                   out.suffix = 'sel.genes.cpd',
                   keys.align = 'y',
                   kegg.native = TRUE,
                   key.pos = demo.paths$kpos1[i],
                   limit = list(gene = 5, cpd = 1),
                   bins = list(gene = 5, cpd = 10),
                   na.col = 'gray',
                   discrete = list(gene = TRUE, cpd = F))

###############################
# ID mapping
###############################

# A distinguished feature of pathview is its strong ID mapping capability. 

# The integrated Mapper module maps over 10 types of gene or protein IDs
# and 20 types of compound or metabolite IDs to standard KEGG gene or 
# compound IDs, and also maps between these external IDs.


cpd.cas <- sim.mol.data(mol.type = 'cpd',
                        id.type = cpd.simtypes[2],
                        nmol = 10000)
head(cpd.cas)

gene.ensport <- sim.mol.data(mol.type = 'gene', 
                             id.type = gene.idtype.list[4],
                             nmol = 50000)
head(gene.ensport)
pv.out <- pathview(gene.data = gene.ensport,
                   cpd.data = cpd.cas,
                   gene.idtype = gene.idtype.list[4],
                   cpd.idtype = cpd.simtypes[2],
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa',
                   same.layer = TRUE,
                   out.suffix = 'gene.ensport.cpd.cas',
                   keys.align = 'x',
                   kegg.native = TRUE,
                   key.pos = demo.paths$kpos2[i],
                   sign.pos = demo.paths$spos[i],
                   limit = list(gene = 3, cpd = 3),
                   bins = list(gene = 6, cpd = 6))

id.map.cas <- cpdidmap(in.ids = names(cpd.cas),
                       in.type = cpd.simtypes[2],
                       out.type = 'KEGG COMPOUND accession')
head(id.map.cas)

cpd.kc <- mol.sum(mol.data = cpd.cas, 
                  id.map = id.map.cas)
head(cpd.kc)

id.map.ensport <- id2eg(ids = names(gene.ensport),
                        category = gene.idtype.list[4],
                        org = 'Hs')
head(id.map.ensport)

gene.entrez <- mol.sum(mol.data = gene.ensport,
                       id.map = id.map.ensport)
head(gene.entrez)

pv.out <- pathview(gene.data = gene.entrez,
                   cpd.data = cpd.kc,
                   pathway.id = demo.paths$sel.paths[i],
                   species = 'hsa',
                   same.layer = 'TRUE',
                   out.suffix = 'gene.entrez.cpd.kc',
                   keys.algin = 'y',
                   kegg.native = TRUE,
                   key.pos = demo.paths$kpos2[i],
                   sign.pos = demo.paths$spos[i],
                   limit = list(gene = 3, cpd = 3),
                   bin = list(gene = 6, cpd = 6))

###################################
# Working with Species
###################################

# KEGG use NCBI GeneID (or Entrez Gene) as the default ID for the most
# common model animals, including human, mouse, rat etc and a few crops,
# e.g. soybean, wine grape and maize. On the other hand, KEGG uses Locus
# tag and other IDs for all others organism, including animals, plants, 
# fungi, protists, as well as bacteria and archaea.

data(korg)
head(korg)

# number of species which use Entrez Gene as the default ID.
sum(korg[, 'entrez.gnodes'] == '1', na.rm = TRUE)

sum(korg[, 'entrez.gnodes'] == '0', na.rm = TRUE)

na.idx <- is.na(korg[, 'ncbi.geneid'])
sum(na.idx)

data(bods)
bods

data("gene.idtype.list")
gene.idtype.list


eco.dat.kegg <- sim.mol.data(mol.type = 'gene',
                             id.type = 'kegg',
                             species = 'eco',
                             nmol = 3000)
head(eco.dat.kegg)

pv.out <- pathview(gene.data = eco.dat.kegg,
                   gene.idtype = 'kegg',
                   pathway.id = '00640',
                   species = 'eco',
                   out.suffix = 'eco.kegg',
                   kegg.native = TRUE,
                   same.layer = TRUE)


eco.dat.entrez <- sim.mol.data(mol.type = 'gene', 
                               id.type = 'entrez',
                               species = 'eco',
                               nmol = 3000)
head(eco.dat.entrez)

pv.out <- pathview(gene.data = eco.dat.entrez, 
                   gene.idtype = 'entrez',
                   pathway.id = '00640',
                   species = 'eco',
                   out.suffix = 'eco.entrez',
                   kegg.native = TRUE)



egid.eco <- eg2id(names(eco.dat.entrez), 
              category = 'symbol',
              pkg = 'org.EcK12.eg.db')
head(egid.eco)

eco.dat.symbol <- eco.dat.entrez
names(eco.dat.symbol) <- egid.eco[, 2]
head(eco.dat.kegg)

pv.out <- pathview(gene.data = eco.dat.symbol, 
                   gene.idtype = 'symbol',
                   pathway.id = '00640',
                   species = 'eco',
                   out.suffix = 'eco.symbol.2layer',
                   kegg.native = TRUE,
                   same.layer = FALSE)


ko.data <- sim.mol.data(mol.type = 'gene.ko',
                        nmol = 5000)
pv.out <- pathview(gene.data = ko.data,
                   pathway.id = '04112',
                   species = 'ko',
                   out.suffix = 'ko.data',
                   kegg.native = TRUE)

############################################
# Integrated workflow with pathway analysis
############################################

library(gage)
data(gse16873)
cn <- colnames(gse16873)
hn <- grep(pattern = 'HN', x = cn, ignore.case = TRUE)
dcis <- grep(pattern = 'DCIS', x = cn, ignore.case = TRUE)
data("kegg.gs")

# pathway analysis using gage
gse16873.kegg.p <- gage(exprs = gse16873, gsets = kegg.gs, 
                        ref = hn, samp = dcis)
head(gse16873.kegg.p$stats, n = 1)
head(gse16873.kegg.p$greater, n = 1)
head(gse16873.kegg.p$less, n = 1)
# prepare the differential expression data
gse16873.d <- gagePrep(exprs = gse16873, ref = hn, 
                       samp = dcis)
head(gse16873.d)

# equivalently, you can do simple subtraction for paired samples.
gse16873.d <- gse16873[, dcis] - gse16873[, hn]
head(gse16873.d)

# select significant pathways and extract their IDs.
sel <- gse16873.kegg.p$greater[, 'q.val'] < 0.1 & !is.na(gse16873.kegg.p$greater[, 'q.val'])
head(sel)

path.ids <- rownames(gse16873.kegg.p$greater)[sel]
head(path.ids)
path.ids2 <- substr(x = path.ids, start = 1, stop = 8)
head(path.ids2)

# pathview visulization 
pv.out.list <- sapply(X = path.ids2, FUN = function(pid) pathview(gene.data = gse16873.d[, 1:2], pathway.id = pid, species = 'hsa'))





