---
title: "DC_RNA-seq"
author: "yincy"
date: "11/30/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
```


```{r}
library(readxl)
library(magrittr)
library(tidyverse)
library(edgeR)
library(biomaRt)
library(pheatmap)
library(factoextra)
library(RColorBrewer)
```


```{r}
dc <- read_xlsx(path = "reads-counts.xlsx")
dc <- dc[, -1]
colnames(dc) <- c("Symbol", "ATP_3um", "ATP_3um+inhibitor", "ATP_500um", "control")
```


```{r filtering}
dc[, 2:5] <- apply(dc[, 2:5], 2, FUN = function(x){(x * 1e6) / colSums(dc[, -1])})

index <- c()
for(i in 1:dim(dc)[1]){
  index <- c(index, sum(dc[i, 2:5] >= 1) >= 2)
}
table(index)
```

```{r}
dc <- dc[index, ]
dc
```



id conversion  
```{r}
mmart <- useEnsembl(biomart = "ensembl", 
                    dataset = "mmusculus_gene_ensembl", 
                    mirror = "asia")
```


```{r}
genes <- getBM(attributes = c("ensembl_gene_id", "entrezgene_id", "mgi_symbol"),
               filters = "mgi_symbol",
               values = dc[, 1, drop = T],
               mart = mmart)
```


```{r}
dc <- left_join(dc, genes, by = c("Symbol" = "mgi_symbol"))
dc <- dc %>% 
  dplyr::select(ensembl_gene_id, entrezgene_id, Symbol, starts_with('ATP'), control)
dc
```


```{r}
hc.dc <- hclust(d = get_dist(x = t(dc[, 4:7]), method = "spearman"), method = "ward.D2")
  
fviz_dend(x = hc.dc,
          horiz = T,
          ggtheme = theme_classic(),
          lwd = 1,
          cex = 0.8,
          main = "Cluster Distance",
          ylab = "distance")
```



## DE genes  
```{r}
dc_mutate <- dc %>% 
  mutate(
    atp.3_vs_control_fc = ATP_3um / control,
    apt.500_vs_control_fc = ATP_500um / control,
    atp.3.inh_vs_control_fc = `ATP_3um+inhibitor` / control,
    )
```


### 3uM ATP - Control  
```{r}
atp.500.deg <- dc_mutate %>% 
  filter(apt.500_vs_control_fc >= 2 | apt.500_vs_control_fc <= 0.5) %>% 
  dplyr::select(Symbol, entrezgene_id,ATP_500um, control, apt.500_vs_control_fc) %>% 
  arrange(apt.500_vs_control_fc)
```

```{r}
goana(de = dc_mutate %>% 
        filter(atp.3_vs_control_fc >= 2 | atp.3_vs_control_fc <= 0.5) %>% 
  dplyr::select(Symbol, entrezgene_id, ATP_3um, control, atp.3_vs_control_fc) %>% 
  arrange(atp.3_vs_control_fc) %>% .[, 2, drop = T])
```



```{r}
kegga(de = atp.500.deg$entrezgene_id, species = "Mm")
```













