---
title: "formate KEGG pathways"
author: "yincy"
date: "1/5/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE}
library(KEGGgraph)
```


## parse pathway to data frame  
```{r}
KGML_files <- list.files(path = "f:/git/Data/KEGG/pathways/mouse", 
                         full.names = T, 
                         pattern = ".xml$")

l <- vector(mode = 'list', length = length(KGML_files))

for(i in seq_along(KGML_files)){
    pathway <- do.call(
        what = "parseKGML2DataFrame", 
        args = list(
            file = KGML_files[i], 
            genesOnly = T, 
            expandGene = T, 
            reaction = T
        )
    )
    n <- gsub(".xml", "", basename(KGML_files[i]))
    l[[i]] <- pathway
    names(l)[[i]] <- n
}

df <- data.frame()
for(i in seq_along(l)){
  df <- rbind(df, l[[i]])
}

df %>% head()

df <- df %>% tibble::as_tibble()

df %>% dplyr::pull(type) %>% table()
df %>% dplyr::pull(subtype) %>% table()

df <- df %>% 
  dplyr::distinct(from, to, .keep_all = T)


df <- df %>% 
  dplyr::filter(type != "maplink")
df %>% dim()

all_path_df <- df

all_path_df$from <- gsub("mmu:", replacement = "", all_path_df$from)
all_path_df$to <- gsub("mmu:", replacement = "", all_path_df$to)

all_path_df %>% dim()
all_path_df %>% head()
```

```{r}
all_path_df %>% 
  dplyr::filter(from == "56543" | to == "56543")
```


## add subtype information  
```{r}
subtype_element <- data.frame(
  name = c("compound", "hidden compound", "activation", "inhibition", 
           "expression", "repression", "indirect effect", "state change", 
           "binding/association", "dissociation", "missing interaction", 
           "phosphorylation", "dephosphorylation", "glycosylation", "ubiquitination", 
           "methylation"), 
  value = c(NA, 
            NA, 
            "-->", "--|", "-->", "--|", "..>", 
            "...", "---", "-+-", "-/-",
            "+p", "-p", "+g", "+u", "+m"), 
  stringsAsFactors = F
)

subtype_element

vl <- subtype_element$value
names(vl) <- subtype_element$name

all_path_df$subtype %>% unique()
intersect(all_path_df$subtype %>% unique(), subtype_element$name)

all_path_df$subtype_value <- vl[all_path_df$subtype]
all_path_df %>% dplyr::filter(subtype != "compound")

## add edge and node info
all_path_df$angle <- ifelse(all_path_df$subtype == "inhibition", 90, 30)
all_path_df$edge_label <- ifelse(grepl("[a-z]", all_path_df$subtype_value), all_path_df$subtype_value, NA)

lty <- c(rep(1, 5), 5, 1, 3, 1, 5, 3, 5, 1, 3, 1)
names(lty) <- all_path_df$subtype %>% unique() %>%  as.character()
all_path_df$lty <- lty[all_path_df$subtype]
all_path_df$lty <- ifelse(is.na(all_path_df$lty), 1, all_path_df$lty)
all_path_df$edge_label <- ifelse(is.na(all_path_df$edge_label), "", all_path_df$edge_label)

all_path_df %>% dplyr::filter(is.na(lty))
```



## add gene and compound information  
```{r}
# get KEGG compound information
KEGG_compounds <- keggList("compound")
KEGG_compounds[10:20]
saveRDS(object = KEGG_compounds, file = "KEGG_compounds.rds")
KEGG_compounds <- readRDS(file = "KEGG_compounds.rds")
KEGG_compound_ids <- names(KEGG_compounds)

pat_location <- str_locate(string = KEGG_compounds, pattern = ";")
ll <- pat_location[, 1]

KEGG_compounds_value <- as.vector(KEGG_compounds)
KEGG_compounds_value[10:20]

extracted_str <- c()
for(i in seq_along(ll)){
  if(is.na(ll[i])){
    extracted_str <- append(extracted_str, KEGG_compounds_value[i])
  }else{
    extracted_str <- append(extracted_str, substr(KEGG_compounds_value[i], start = 1, stop = ll[i] - 1))
  }
}

KEGG_compounds_to_display <- str_trim(extracted_str, "both")
names(KEGG_compounds_to_display) <- KEGG_compound_ids
KEGG_compounds_to_display[10:20]
```

```{r}
# get KEGG gene information
KEGG_mmu_genes <- keggList("mmu")
saveRDS(object = mmu_genes, file = "KEGG_mmu_genes.rds")
KEGG_mmu_genes <- readRDS(file = "KEGG_mmu_genes.rds")
KEGG_mmu_genes %>% head()

library(org.Mm.eg.db)
KEGG_mmu_genes_df <- select(x = org.Mm.eg.db, 
                       keys = KEGG_mmu_genes, 
                       keytype = "ENTREZID", 
                       columns = c("ENTREZID", "SYMBOL"))
KEGG_mmu_genes_df

KEGG_mmu_genes_to_display <- KEGG_mmu_genes_df$SYMBOL
names(KEGG_mmu_genes_to_display) <- KEGG_mmu_genes_df$ENTREZID
KEGG_mmu_genes_to_display %>% head()
```

```{r}
# get KEGG glycan information  
KEGG_glycan <- keggList("glycan")
saveRDS(object = KEGG_glycan, file = "KEGG_glycan.rds")
KEGG_glycan <- readRDS(file = "KEGG_glycan.rds")
KEGG_glycan %>% head()

KEGG_glycan_ids <- names(KEGG_glycan)
KEGG_glycan_ids %>% head()

KEGG_glycan_values <- KEGG_glycan %>% as.vector()
KEGG_glycan_values %>% head()

pat_location <- str_locate(KEGG_glycan, ";")
pat_location %>% dim()
ll <- pat_location[, 1]

extracted_str <- c()
for(i in seq_along(ll)){
  if(is.na(ll[i])){
    extracted_str <- append(extracted_str, KEGG_glycan_values[i])
  }else{
    extracted_str <- append(extracted_str, substr(KEGG_glycan_values[i], start = 1, stop = ll[i] - 1))
  }
}

KEGG_glycan_to_dispaly <- str_trim(extracted_str, "both")
names(KEGG_glycan_to_dispaly) <- KEGG_glycan_ids
KEGG_glycan_to_dispaly[1:10]
```


```{r}
# get KEGG drug information 
KEGG_drug <- keggList("drug")
saveRDS(object = KEGG_drug, file = "KEGG_drug.rds")
KEGG_drug <- readRDS(file = "KEGG_drug.rds")
KEGG_drug %>% head()

KEGG_drug_ids <- names(KEGG_drug)
KEGG_drug_values <- KEGG_drug %>% as.vector()
KEGG_drug_values %>% head()

pat_location <- str_locate(string = KEGG_drug, pattern = ";")
ll <- pat_location[, 1]

extracted_str <- c()
for(i in seq_along(ll)){
  if(is.na(ll[i])){
    extracted_str <- append(extracted_str, KEGG_drug_values[i])
  }else{
    extracted_str <- append(extracted_str, substr(KEGG_drug_values[i], start = 1, stop = ll[i] - 1))
  }
}

KEGG_drug_to_display <- extracted_str

## remove parenthesis 
KEGG_drug_to_display <- str_replace(KEGG_drug_to_display, "[ ]*\\(?[[:alnum:][:punct:]]*\\)?$", "")
names(KEGG_drug_to_display) <- KEGG_drug_ids
KEGG_drug_to_display %>% head()
```


```{r}
# some rows in from and to contain multiple ids, retain only the first one
ll <- str_locate(all_path_df$from, " ") %>% .[, 1]
all_path_df[!(ll %>% is.na()), ]

for(i in seq_along(ll)){
  if(is.na(ll[i])){
    all_path_df$from[i] <- all_path_df$from[i]
  }else{
    all_path_df$from[i] <- str_extract_before(all_path_df$from[i], " ")
  }
}


ll <- str_locate(all_path_df$to, " ") %>% .[, 1]
all_path_df[!(ll %>% is.na()), ]

for(i in seq_along(ll)){
  if(is.na(ll[i])){
    all_path_df$to[i] <- all_path_df$to[i]
  }else{
    all_path_df$to[i] <- str_extract_before(all_path_df$to[i], " ")
  }
}

all_path_df$from <- str_trim(string = all_path_df$from, side = "both")
all_path_df$to <- str_trim(string = all_path_df$to, side = "both")

names_to_display <- c(KEGG_mmu_genes_to_display, KEGG_compounds_to_display, 
                      KEGG_drug_to_display, KEGG_glycan_to_dispaly)

# all_path_df$from_symbol <- names_to_display[all_path_df$from]
# all_path_df$to_symbol <- names_to_display[all_path_df$to]
# all_path_df <- all_path_df %>% 
#   dplyr::rename(from = from_symbol, to = to_symbol, from_entrez = from, to_entrez = to)
# all_path_df %>% dplyr::filter(is.na(from)|is.na(to))
```


**Type attribute**  
attribute value| explanation
---------------|-----------------------
ECrel| enzyme-enzyme relation, indicating two enzyme catalyzing successive reaction steps  
PPrel| protein-protein interaction, such as binding and modification
GErel| gene expression interaction, indicating relation of transcription factor and target gene product
PCrel| protein-compound interaction 
maplink| link to another map


```{r}
saveRDS(object = all_path_df, file = "../data/KEGG/KEGG_pathways_formatted.rds")
saveRDS(object = names_to_display, file = "../data/KEGG/names_to_display.rds")
```

