---
title: "P2y12r and C1q"
author: "yincy"
date: "9/14/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r, message=FALSE, warning=FALSE, echo=FALSE}
library(tidyverse)
library(EnsDb.Mmusculus.v79)
library(KEGGREST)
library(pathview)
library(KEGGgraph)
library(biomaRt)
library(igraph)
```


# P2y12r signaling network  
## PVN  
```{r}
pvn <- read.csv(file = "../data/20200702_frame_tpm_coding_gene_PVN.csv", 
                row.names = 1)
pvn
```

samples  
```{r}
A14 <- grep("14A", colnames(pvn), value = T)
L14 <- grep("14L", colnames(pvn), value = T)
A3 <- grep("3A", colnames(pvn), value = T)
L3 <- grep("3L", colnames(pvn), value = T)
naive <- grep("naive", colnames(pvn), value = T)
```

fold changes  
```{r}
FC_df <- pvn %>% 
    rownames_to_column(var = "ensembl") %>% 
    rowwise() %>% 
    mutate(
        mean_A14 = mean(PVN_14A1, PVN_14A2, PVN_14A3), 
        mean_L14 = mean(PVN_14L1, PVN_14L2_b1, PVN_14L3_b1, PVN_14L4, PVN_14L5, PVN_14L6), 
        mean_A3 = mean(PVN_3A1_b1, PVN_3A2, PVN_3A3_b1), 
        mean_L3 = mean(PVN_3L1_b1, PVN_3L2_b1, PVN_3L3), 
        mean_naive = mean(PVN_naive1, PVN_naive2, PVN_naive3), 
        FC_A14 = mean_A14 / mean_naive, 
        FC_L14 = mean_L14 / mean_naive, 
        FC_A3 = mean_A3 / mean_naive, 
        FC_L3 = mean_L3 / mean_naive) %>% 
    relocate(geneid_symbol, ensembl, starts_with("FC_"), starts_with("mean")) 

FC_df[!duplicated(FC_df$geneid_symbol), ]
```

```{r}
agg_FC_df <- aggregate.data.frame(FC_df[, 3:6], list(FC_df$geneid_symbol), FUN = "sum") %>% 
    column_to_rownames("Group.1")
agg_FC_df
```

```{r}
agg_FC_df_log2 <- agg_FC_df %>% log2()

for(i in dim(agg_FC_df_log2)[1]){
    for(j in dim(agg_FC_df_log2)[2]){
        ifelse(agg_FC_df_log2[i, j] == -Inf, -5, agg_FC_df_log2[i, j])
        ifelse(agg_FC_df_log2[i, j] == Inf, 5, agg_FC_df_log2[i, j])
        ifelse(is.na(agg_FC_df_log2[i, j]), 0, agg_FC_df_log2[i, j])
        ifelse(abs(agg_FC_df_log2[i, j]) > 5, 5 * sign(agg_FC_df_log2[i, j]), agg_FC_df_log2[i, j])
    }
}

agg_FC_df_log2
```

```{r}
par(mfrow = c(3, 1))
plot(agg_FC_df_log2$FC_A14, agg_FC_df$FC_L14, pch = 19, xlim = c(-9, 9), ylim = c(-9, 9))
hist(agg_FC_df_log2$FC_A14, breaks = 30)
hist(agg_FC_df_log2$FC_L14, breaks = 30)
```


## map p2y12r onto KEGG pathways  
```{r}
p2ry12_entrez <- "70839"
mgs[grep("p2ry12", mgs$gene_name, ignore.case = T, value = F)]
```

```{r}
keggLink(target = "pathway", source = paste("mmu:", p2ry12_entrez, sep = ""))
```

```{r}
pathview(gene.data = agg_FC_df, 
         pathway.id = "04611", 
         kegg.native = F, 
         species = "mmu", 
         gene.idtype = "SYMBOL", 
         cpd.lab.offset = 0, 
         cex = 0.4, 
         min.nnodes = 1, 
         trans.fun = list(gene = log2), 
         multi.state =F, 
         limit = list(gene = c(-5, 5), cpd = c(-5, 5)))
```


## parse pathway to data frame   
```{r}
KGML_files <- list.files(path = "/home/yincy/git/Data/KEGG/pathways/mouse/", 
                         full.names = T, 
                         pattern = ".xml$")
# KGML_files <- KGML_files[file.size(KGML_files) < 2300000]

df <- data.frame()
for(i in KGML_files){
    pathway <- do.call(
        what = "parseKGML2DataFrame", 
        args = list(
            file = i, 
            genesOnly = T, 
            expandGene = T
        )
    )
    df <- rbind(df, pathway)
}

df <- df %>% as_tibble()
write.csv(x = df, file = "all_path_dataframe.csv", 
          row.names = F, 
          quote = F)

all_path_df <- read.csv(file = "all_path_dataframe.csv", 
                        stringsAsFactors = F)
all_path_df %>% dim()
all_path_df %>% pull(type) %>% table()
all_path_df %>% pull(subtype) %>% table()

all_path_df %>% head()
all_path_df <- all_path_df %>% distinct(from, to, .keep_all = T)
all_path_df %>% dim()

from <- all_path_df %>% pull(from)
to <- all_path_df %>% pull(to)
c(from, to) %>% unique() %>% length()
```

```{r}
all_path_df$from <- gsub("mmu:", "", all_path_df$from)
all_path_df$to <- gsub("mmu:", "", all_path_df$to)
all_path_df
```

```{r}
mmart <- useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
from_df <- getBM(attributes = c("entrezgene_id", "mgi_symbol"), 
                 filters = "entrezgene_id", 
                 values = all_path_df[, "from"] %>% unique(), 
                 mart = mmart)
from_df <- from_df %>% 
    distinct(entrezgene_id, .keep_all = T) %>% 
    column_to_rownames("entrezgene_id") 

from_vector <- from_df[[1]]
names(from_vector) <- rownames(from_df)
from_vector %>% length()
from_vector %>% head()


to_df <- getBM(attributes = c("entrezgene_id", "mgi_symbol"), 
                 filters = "entrezgene_id", 
                 values = all_path_df[, "to"] %>% unique(), 
                 mart = mmart)
to_df <- to_df %>% 
    distinct(entrezgene_id, .keep_all = T) %>% 
    column_to_rownames("entrezgene_id")

to_vector <- to_df[[1]]
names(to_vector) <- rownames(to_df)
to_vector %>% length()
to_vector %>% head()
```

```{r}
all_path_df$from_symbol <- from_vector[all_path_df$from] 
all_path_df$to_symbol <- to_vector[all_path_df$to]
all_path_df <- all_path_df %>% 
    dplyr::select(from_symbol, to_symbol, from, to, type, subtype)
```

```{r}
all_path_df %>% pull(from_symbol) %>% is.na() %>% table()
all_path_df %>% pull(to_symbol) %>% is.na() %>% table()
```

```{r}
all_path_graph <- graph_from_data_frame(d = all_path_df, 
                                        directed = T)

ego_size(graph = all_path_graph, order = 3, nodes = "P2ry12", mode = "out")
p2ry12_order1 <- make_ego_graph(graph = all_path_graph, order = 1, nodes = "P2ry12", mode = "out")[[1]]
p2ry12_order2 <- make_ego_graph(graph = all_path_graph, order = 2, nodes = "P2ry12", mode = "out")[[1]]
p2ry12_order3 <- make_ego_graph(graph = all_path_graph, order = 3, nodes = "P2ry12", mode = "out")[[1]]

nodes_o1 <- p2ry12_order1 %>% 
    as_data_frame() %>% 
    .[,2] %>% 
    unique()

nodes_o2 <- p2ry12_order2 %>% 
    as_data_frame() %>% 
    .[, 2] %>% 
    unique() %>% 
    setdiff(nodes_o1)

nodes_o3 <- p2ry12_order3 %>% 
    as_data_frame() %>% 
    .[, 2] %>% 
    unique() %>% 
    setdiff(nodes_o1) %>% 
    setdiff(nodes_o2)

p2ry12_o1 <- p2ry12_order3 %>% 
    as_data_frame() %>% 
    magrittr::set_colnames(c("from_symbol", "to_symbol", "from_entrez", "to_entrez", "type", "subtype")) %>% 
    dplyr::filter(from_symbol %in% c("P2ry12")) %>% 
    distinct(from_symbol, to_symbol, .keep_all = T)

p2ryo2 <- p2ry12_order3 %>% 
    as_data_frame() %>% 
    magrittr::set_colnames(c("from_symbol", "to_symbol", "from_entrez", "to_entrez", "type", "subtype")) %>% 
    dplyr::filter(from_symbol %in% nodes_o1) %>% 
    distinct(from_symbol, to_symbol, .keep_all = T)

p2ryo3 <- p2ry12_order3 %>% 
    as_data_frame() %>% 
    magrittr::set_colnames(c("from_symbol", "to_symbol", "from_entrez", "to_entrez", "type", "subtype")) %>% 
    dplyr::filter(from_symbol %in% nodes_o2) %>% 
    distinct(from_symbol, to_symbol, .keep_all = T)

p2ry12_sig_df <- rbind(p2ry12_o1, p2ryo2, p2ryo3) 

p2ry12_ig <- graph_from_data_frame(d = p2ry12_sig_df, directed = T)

vcol <- rep("grey70", V(p2ry12_ig) %>% length())
vcol[V(p2ry12_ig)$name == "P2ry12"] <- "red"
vcol[V(p2ry12_ig)$name %in% nodes_o1] <- "gold"
vcol[V(p2ry12_ig)$name %in% nodes_o2] <- "blue"
vcol %>% table()

ecol <- rep("grey70", E(p2ry12_ig) %>% length())
table(ecol)

ll <- layout_with_fr(graph = p2ry12_ig)
ll %>% head()
ll %>% .[, 1] %>% range()
ll %>% .[, 2] %>% range()

pdf(file = "P2ry12_order3.pdf")
plot(p2ry12_ig, 
     vertex.size = 15, 
     vertex.color = vcol, 
     vertex.label.cex = 0.3,
     vertex.label.color = vcol,
     vertex.frame.color = NA, 
     edge.arrow.size = 0.2, 
     xlim = ll %>% .[, 1] %>% range(), 
     ylim = ll %>% .[, 2] %>% range(), 
     rescale = F, 
     layout = ll*1.1, 
#     vertex.label = NA, 
     edge.color = ecol, 
     edge.width = 0.2, 
    edge.curved = 0.2)

legend("topright", 
       legend = c("order 0 = 1", "order 1 = 3", "order 2 = 86", "order 3 = 559"), 
       pch = 19, 
       col = c("red", "gold", "blue", "grey70"), 
       bty = "n", 
       cex = 0.7)

dev.off()
```

```{r}
write.csv(x = p2ry12_sig_df[, 1:2] %>% distinct(from_symbol, to_symbol), 
          file = "p2ry12_order3_df.csv", 
          row.names = F, 
          quote = F)
```


## DEG in p2ry12 order 3 genes  
```{r}
A3 <- read.csv(file = "../data/20200702_3A_naive_DE_FDR.csv", header = T)
A3_DEG <- A3 %>% dplyr::filter(abs(logFC) >= 1) %>% 
    dplyr::pull(geneid_symbol)

L3 <- read.csv(file = "../data/20200702_3L_naive_DE_FDR.csv", header = T)
L3_DEG <- L3 %>% 
    dplyr::filter(abs(logFC) >= 1) %>% 
    dplyr::pull(geneid_symbol)
```


```{r}
A14 <- read.csv(file = "../data/20200702_14A_naive_DE_FDR.csv")
A14_DEG <- A14 %>% 
  dplyr::filter(abs(logFC) >= 1) %>% 
  dplyr::pull(geneid_symbol)

L14 <- read.csv(file = "../data/20200702_14L_naive_DE_FDR.csv")
L14_DEG <- L14 %>%
  dplyr::filter(abs(logFC) >= 1) %>% 
  dplyr::pull(geneid_symbol)
```


```{r}
p2ry12_order3_sig_df <- read.csv(file = "p2ry12_order3_df.csv", header = T, stringsAsFactors = F)

from_to_vector <- c(p2ry12_order3_sig_df[[1]], p2ry12_order3_sig_df[[2]]) %>% unique()
```

```{r}
A3_DEG_order3 <- intersect(from_to_vector, A3_DEG)
L3_DEG_order3 <- intersect(from_to_vector, L3_DEG)
```


```{r}
write.csv(x = A3_DEG_order3, 
          file = "A3_DEG_order3_intersect.csv", 
          row.names = F, 
          quote = F)

write.csv(x = L3_DEG_order3, 
          file = "L3_DEG_order3_intersect.csv", 
          row.names = F, 
          quote = F)
```

```{r}
intersects <- intersect(L3_DEG_order3, A3_DEG_order3)
```


## GO  
```{r}
go_list_pvn_2 <- readxl::read_xlsx(path = "GO list/20200812_geneid_updown_go_list_PVN_uniq.xlsx", 
                                   sheet = 1)

go_list_pvn_2 %>% 
    as_tibble() %>% 
    dplyr::arrange(pvalue) %>% 
    head(n = 100) %>% 
    pull(geneID) %>% 
    str_c(collapse = "/") %>% 
    str_split(pattern = "/") %>% 
    .[[1]] %>% 
    unique() 
```


```{r}
files <- list.files(path = "GO list", pattern = "BS_|CTX_", full.names = T)
sheets <- c("3up_GO", "3down_GO", "co_up_GO", "co_down_GO")
file_names <- c("BS_2.xlsx", "BS_uniq.xlsx", "CTX_2.xlsx", "CTX_uniq.xlsx")

for(i in files[4]){
    for(sheet in sheets){
        p <- read.xlsx(file = i, sheetName = sheet)
        p$overlap <- p$geneID %>% str_split(pattern = "/") %>% 
            lapply(FUN = intersect, y = intersects) %>% 
            sapply(FUN = str_c, collapse = "/")
        write.xlsx(x = p, file = "CTX_uniq.xlsx", append = T, sheetName = sheet)
    }
}
```

## top 100 GO term genes  
```{r}
brain_region <- rep(c("BS", "CTX", "PVN"), each = 8)
file_index <- rep(rep(c("2", "uniq"), each = 4), time = 3)
sheets <- rep(c("3up_GO", "3down_GO", "co_up_GO", "co_down_GO"), time = 6)
files <- paste("GO list/", list.files(path = "GO list", pattern = "_PVN_|_BS_|_CTX_"), sep = "") %>% rep(each = 4)

out <- paste(brain_region, file_index, sheets, sep = "_")

out_list <- list()
len <- c()

for(i in seq_along(out)){
    x <- xlsx::read.xlsx(file = files[i], sheetName = sheets[i]) %>% 
        tibble::as_tibble() %>% 
        dplyr::arrange(pvalue) %>% 
        head(n = 100) %>% 
        dplyr::pull(geneID) %>% 
        stringr::str_c(collapse = "/") %>% 
        stringr::str_split(pattern = "/") %>% 
        .[[1]] %>% 
        unique()
    y <- length(x)
    out_list[[i]] <- x
}
```

```{r}
df <- out_list %>% 
    unlist() %>% 
    as.data.frame() %>% 
    tibble::rownames_to_column("group") %>% 
    magrittr::set_colnames(c("group", "symbol"))

df$sample <- gsub(pattern = "[0-9]{,2}$", replacement = "", x = df$sample)

df %>% 
    group_by(sample) %>% 
    summarise(n_genes = n()) %>% 
    arrange(sample) %>% 
    write.csv(file = "number_of_unique_genes_per_group.csv", 
              row.names = F, 
              quote = F)
```

```{r}
df %>% 
    write.csv(file = "genes_per_group.csv", row.names = F, quote = F)
```


# PVN DGE in top100 GO DGE map to KEGG pathway  
## get intersects genes  
```{r}
pvn_top_100_GO_DEG <- read.csv(file = "../res/P2ry12/genes_per_group.csv") %>% 
    dplyr::filter(group %in% c("PVN_2_3up_GO", "PVN_2_3down_GO")) %>% 
    .[[2]] %>% 
    unique()

A3_DEG_order3_intersect <- read.csv(file = "../res/P2ry12/A3_DEG_order3_intersect.csv",
                                    header = F) %>% .[[1]]

L3_DEG_order3_intersect <- read.csv(file = "../res/P2ry12/L3_DEG_order3_intersect.csv", 
                                    header = F) %>% .[[1]] 

intersects_A3_L3 <- intersect(A3_DEG_order3_intersect, L3_DEG_order3_intersect)

intersects_pvn_top100_DEG_A3_L3_DGE <- intersect(pvn_top_100_GO_DEG, intersects_A3_L3)
```


## parse pathway to data frame  
```{r}
KGML_files <- list.files(path = "/home/yincy/git/Data/KEGG/pathways/mouse", 
                         full.names = T, 
                         pattern = ".xml$")

df <- data.frame()
for(i in KGML_files){
    pathway <- do.call(
        what = "parseKGML2DataFrame", 
        args = list(
            file = i, 
            genesOnly = F, 
            expandGene = T, 
            reaction = T
        )
    )
    df <- rbind(df, pathway)
}

df <- df %>% tibble::as_tibble()
write.csv(x = df, 
          file = "../res/P2ry12/all_pathways_df.csv", 
          row.names = F, 
          quote = F)

all_path_df <- read.csv(file = "../res/P2ry12/all_pathways_df.csv", 
                        stringsAsFactors = F)

all_path_df %>% dim()
all_path_df %>% pull(type) %>% table()
all_path_df %>% pull(subtype) %>% table()

all_path_df %>% head()
all_path_df <- all_path_df %>% distinct(from, to, .keep_all = T)
all_path_df %>% dim()

all_path_df %>% 
  dplyr::filter(type == "maplink")

all_path_df <- all_path_df %>% 
  dplyr::filter(type != "maplink")
all_path_df %>% dim()

all_path_df$from <- gsub("mmu:", replacement = "", all_path_df$from)
all_path_df$to <- gsub("mmu:", replacement = "", all_path_df$to)

all_path_df %>% dim()
all_path_df %>% head()
```

## add subtype information  
```{r}
subtype_element <- data.frame(
  name = c("compound", "hidden compound", "activation", "inhibition", 
           "expression", "repression", "indirect effect", "state change", 
           "binding/association", "dissociation", "missing interaction", 
           "phosphorylation", "dephosphorylation", "glycosylation", "ubiquitination", 
           "methylation"), 
  value = c(NA, 
            NA, 
            "-->", "--|", "-->", "--|", "..>", 
            "...", "---", "-+-", "-/-",
            "+p", "-p", "+g", "+u", "+m"), 
  stringsAsFactors = F
)

subtype_element

vl <- subtype_element$value
names(vl) <- subtype_element$name

all_path_df$subtype %>% unique()
intersect(all_path_df$subtype %>% unique(), subtype_element$name)

all_path_df$subtype_value <- vl[all_path_df$subtype]
all_path_df %>% dplyr::filter(subtype != "compound")

## add edge and node info
all_path_df$angle <- ifelse(all_path_df$subtype == "inhibition", 90, 30)
all_path_df$edge_label <- ifelse(grepl("[a-z]", all_path_df$subtype_value), all_path_df$subtype_value, NA)

lty <- c(rep(1, 5), 5, 1, 3, 1, 5, 3, 5, 1, 3, 1)
names(lty) <- all_path_df$subtype %>% unique() %>%  as.character()
all_path_df$lty <- lty[all_path_df$subtype]
all_path_df$lty <- ifelse(is.na(all_path_df$lty), 1, all_path_df$lty)
all_path_df$edge_label <- ifelse(is.na(all_path_df$edge_label), "", all_path_df$edge_label)

all_path_df %>% dplyr::filter(is.na(lty))
```


## add gene and compound information  
```{r}
# get KEGG compound information
KEGG_compounds <- keggList("compound")
KEGG_compounds[10:20]
saveRDS(object = KEGG_compounds, file = "KEGG_compounds.rds")
KEGG_compounds <- readRDS(file = "KEGG_compounds.rds")
KEGG_compound_ids <- names(KEGG_compounds)

pat_location <- str_locate(string = KEGG_compounds, pattern = ";")
ll <- pat_location[, 1]

KEGG_compounds_value <- as.vector(KEGG_compounds)
KEGG_compounds_value[10:20]

extracted_str <- c()
for(i in seq_along(ll)){
  if(is.na(ll[i])){
    extracted_str <- append(extracted_str, KEGG_compounds_value[i])
  }else{
    extracted_str <- append(extracted_str, substr(KEGG_compounds_value[i], start = 1, stop = ll[i] - 1))
  }
}

KEGG_compounds_to_display <- str_trim(extracted_str, "both")
names(KEGG_compounds_to_display) <- KEGG_compound_ids
KEGG_compounds_to_display[10:20]
```

```{r}
# get KEGG gene information
KEGG_mmu_genes <- keggList("mmu")
saveRDS(object = mmu_genes, file = "KEGG_mmu_genes.rds")
KEGG_mmu_genes <- readRDS(file = "KEGG_mmu_genes.rds")
KEGG_mmu_genes %>% head()

library(org.Mm.eg.db)
KEGG_mmu_genes_df <- select(x = org.Mm.eg.db, 
                       keys = KEGG_mmu_genes, 
                       keytype = "ENTREZID", 
                       columns = c("ENTREZID", "SYMBOL"))
KEGG_mmu_genes_df

KEGG_mmu_genes_to_display <- KEGG_mmu_genes_df$SYMBOL
names(KEGG_mmu_genes_to_display) <- KEGG_mmu_genes_df$ENTREZID
KEGG_mmu_genes_to_display %>% head()
```

```{r}
# get KEGG glycan information  
KEGG_glycan <- keggList("glycan")
saveRDS(object = KEGG_glycan, file = "KEGG_glycan.rds")
KEGG_glycan <- readRDS(file = "KEGG_glycan.rds")
KEGG_glycan %>% head()

KEGG_glycan_ids <- names(KEGG_glycan)
KEGG_glycan_ids %>% head()

KEGG_glycan_values <- KEGG_glycan %>% as.vector()
KEGG_glycan_values %>% head()

pat_location <- str_locate(KEGG_glycan, ";")
pat_location %>% dim()
ll <- pat_location[, 1]

extracted_str <- c()
for(i in seq_along(ll)){
  if(is.na(ll[i])){
    extracted_str <- append(extracted_str, KEGG_glycan_values[i])
  }else{
    extracted_str <- append(extracted_str, substr(KEGG_glycan_values[i], start = 1, stop = ll[i] - 1))
  }
}

KEGG_glycan_to_dispaly <- str_trim(extracted_str, "both")
names(KEGG_glycan_to_dispaly) <- KEGG_glycan_ids
KEGG_glycan_to_dispaly[1:10]
```


```{r}
# get KEGG drug information 
KEGG_drug <- keggList("drug")
saveRDS(object = KEGG_drug, file = "KEGG_drug.rds")
KEGG_drug <- readRDS(file = "KEGG_drug.rds")
KEGG_drug %>% head()

KEGG_drug_ids <- names(KEGG_drug)
KEGG_drug_values <- KEGG_drug %>% as.vector()
KEGG_drug_values %>% head()

pat_location <- str_locate(string = KEGG_drug, pattern = ";")
ll <- pat_location[, 1]

extracted_str <- c()
for(i in seq_along(ll)){
  if(is.na(ll[i])){
    extracted_str <- append(extracted_str, KEGG_drug_values[i])
  }else{
    extracted_str <- append(extracted_str, substr(KEGG_drug_values[i], start = 1, stop = ll[i] - 1))
  }
}

KEGG_drug_to_display <- extracted_str

## remove parenthesis 
KEGG_drug_to_display <- str_replace(KEGG_drug_to_display, "[ ]*\\(?[[:alnum:][:punct:]]*\\)?$", "")
names(KEGG_drug_to_display) <- KEGG_drug_ids
KEGG_drug_to_display %>% head()
```


```{r}
# some rows in from and to contain multiple ids, retain only the first one
ll <- str_locate(all_path_df$from, " ") %>% .[, 1]
all_path_df[!(ll %>% is.na()), ]

for(i in seq_along(ll)){
  if(is.na(ll[i])){
    all_path_df$from[i] <- all_path_df$from[i]
  }else{
    all_path_df$from[i] <- str_extract_before(all_path_df$from[i], " ")
  }
}


ll <- str_locate(all_path_df$to, " ") %>% .[, 1]
all_path_df[!(ll %>% is.na()), ]

for(i in seq_along(ll)){
  if(is.na(ll[i])){
    all_path_df$to[i] <- all_path_df$to[i]
  }else{
    all_path_df$to[i] <- str_extract_before(all_path_df$to[i], " ")
  }
}

all_path_df$from <- str_trim(string = all_path_df$from, side = "both")
all_path_df$to <- str_trim(string = all_path_df$to, side = "both")

names_to_display <- c(KEGG_mmu_genes_to_display, KEGG_compounds_to_display, 
                      KEGG_drug_to_display, KEGG_glycan_to_dispaly)

# all_path_df$from_symbol <- names_to_display[all_path_df$from]
# all_path_df$to_symbol <- names_to_display[all_path_df$to]
# all_path_df <- all_path_df %>% 
#   dplyr::rename(from = from_symbol, to = to_symbol, from_entrez = from, to_entrez = to)
# all_path_df %>% dplyr::filter(is.na(from)|is.na(to))
```


**Type attribute**  
attribute value| explanation
---------------|-----------------------
ECrel| enzyme-enzyme relation, indicating two enzyme catalyzing successive reaction steps  
PPrel| protein-protein interaction, such as binding and modification
GErel| gene expression interaction, indicating relation of transcription factor and target gene product
PCrel| protein-compound interaction 
maplink| link to another map


## visulization with `ggraph`  

```{r}
library(ggraph)
library(tidygraph)
```

### determine order of genes  
```{r}
# all_path_gene_df <- read.csv(file = "../res/P2ry12/all_path_dataframe.csv") %>% 
#   distinct(from, to, .keep_all = T)
# write.csv(x = all_path_gene_df, file = "../res/P2ry12/all_path_gene_df.csv")

# all_path_gene_df$from <- gsub("mmu:", "", all_path_gene_df$from)
# all_path_gene_df$to <- gsub("mmu:", "", all_path_gene_df$to)
# 
# all_path_gene_df$from_symbol <- KEGG_mmu_genes_to_display[all_path_gene_df$from]
# all_path_gene_df$to_symbol <- KEGG_mmu_genes_to_display[all_path_gene_df$to]
# 
# all_path_gene_df <- all_path_gene_df %>% 
#   relocate(from_symbol, to_symbol, everything())
```


```{r}
all_path_df
```


```{r}
# all_path_df <- all_path_df %>% 
#   as.tibble() %>% 
#   dplyr::rename(from = from_symbol, to = to_symbol, from_entrez = from, to_entrez = to)

ig <- graph_from_data_frame(d = all_path_df[, c("from", "to")], directed = T)
distance_to_p2ry12_all_path <- shortest_paths(ig, from = V(ig)$name == "70839", to = V(ig), mode = "out")[[1]]

list2df <- function(list){
  maxL <- sapply(list, length) %>% max()
  for(i in seq_along(list)){
    l = length(list[[i]])
    if(l < maxL){
      list[[i]] <- append(list[[i]], values = rep(NA, (maxL - l)))
    }
  }
  list <- as.data.frame(list)
  colnames(list) <- NULL
  list <- t(list)
  colnames(list) <- paste(rep("order", times = maxL), 0:(maxL - 1), sep = "_")
  return(as.data.frame(list))
}

all_path_nodes <- list2df(lapply(distance_to_p2ry12_all_path, as_ids))

inter_len <- apply(all_path_nodes[1:10], 
                   1, 
                   intersect, 
                   mapIds(org.Mm.eg.db, keys = all_DEG, keytype = "SYMBOL", column = "ENTREZID")) %>%
  sapply(length)

retained_path <- all_path_nodes[inter_len >= 1, ] 

o1 <- all_path_df %>% 
  filter(from %in% (retained_path %>% pull(order_0) %>% unique()), 
         to %in% (retained_path %>% pull(order_1) %>% unique()))

o2 <- all_path_df %>% 
  filter(from %in% (retained_path %>% pull(order_1) %>% unique()), 
         to %in% (retained_path %>% pull(order_2) %>% unique()))

o3 <- all_path_df %>% 
  filter(from %in% (retained_path %>% pull(order_2) %>% unique()), 
         to %in% (retained_path %>% pull(order_3) %>% unique()))

o4 <- all_path_df %>% 
  filter(from %in% (retained_path %>% pull(order_3) %>% unique()), 
         to %in% (retained_path %>% pull(order_4) %>% unique()))

o5 <- all_path_df %>% 
  filter(from %in% (retained_path %>% pull(order_4) %>% unique()), 
         to %in% (retained_path %>% pull(order_5) %>% unique()))
o6 <- all_path_df %>% 
  filter(from %in% (retained_path %>% pull(order_5) %>% unique()), 
         to %in% (retained_path %>% pull(order_6) %>% unique()))

o7 <- all_path_df %>% 
  filter(from %in% (retained_path %>% pull(order_6) %>% unique()), 
         to %in% (retained_path %>% pull(order_7) %>% unique()))

o8 <- all_path_df %>% 
  filter(from %in% (retained_path %>% pull(order_7) %>% unique()), 
         to %in% (retained_path %>% pull(order_8) %>% unique()))


filtered_path <- rbind(o1, o2, o3, o4, o5, o6, o7, o8)
filtered_path %>% pull(lty) 

N <- tibble::tibble(
  ids = c(filtered_path$from, filtered_path$to) %>% unique()
) %>% 
  mutate(shape = ifelse(grepl("^cpd:|^gl:|^dr:", ids, ignore.case = F), "21", "22")) 

E <- filtered_path %>% 
  dplyr::select(from, to, everything())
```


DEG   
```{r}
A3_DEG_order3_intersect <- read.csv(file = "../res/P2ry12/A3_DEG_order3_intersect.csv")[[1]]
L3_DEG_order3_intersect <- read.csv(file = "../res/P2ry12/L3_DEG_order3_intersect.csv")[[1]]
A3_L3_DEG_inter <- intersect(A3_DEG_order3_intersect, L3_DEG_order3_intersect)

pvn_top100_GO_DEG <- read.csv(file = "../res/P2ry12/genes_per_group.csv") %>% 
  filter(grepl("^PVN_2_3up|^PVN_2_3down", group)) %>% 
  pull(symbol)

write.csv(x = pvn_top100_GO_DEG, file = "../res/P2ry12/pvn_top100_GO_DEG.csv")

intersect(A3_DEG_order3_intersect, pvn_top100_GO_DEG)
intersect(L3_DEG_order3_intersect, pvn_top100_GO_DEG)
all_DEG <- intersect(A3_L3_DEG_inter, pvn_top100_GO_DEG)
```

```{r, warning=FALSE, message=FALSE}
g <- tbl_graph(nodes = N, edges = E, directed = T)

# ll <- create_layout(g, "nicely")
set.seed(123)
g %>%  
  mutate(names = names_to_display[ids]) %>% 
  ggraph(layout = "lgl") +
  geom_edge_link(aes(filter = (angle == 30), 
                     edge_linetype = factor(lty)), 
                 end_cap = circle(2.5, "mm"), 
                 start_cap = circle(1, "mm"), 
                 arrow = arrow(length = unit(1, "mm")), 
                 show.legend = F, 
                 edge_width = 0.2) +
  geom_edge_link(aes(filter = (angle == 90), 
                     edge_linetype = factor(lty)), 
                 arrow = arrow(angle = 90, length = unit(1, "mm")), 
                 end_cap = circle(2.5, "mm"), 
                 show.legend = F, 
                 edge_width = 0.2) +
  geom_node_point(aes(shape = shape, filter = !(names %in% A3_L3_DEG_inter)), 
                  fill ="grey", 
                  size = 5, 
                  show.legend = F, 
                  color = "grey") +
  geom_node_point(aes(shape = shape, filter = (names %in% A3_L3_DEG_inter)), 
             fill = "red", 
             color = "red", 
             show.legend = F, 
             size = 5) +
  geom_node_text(aes(label = names), size = 2) +
  scale_shape_manual(values = c("21" = 22, "22" = 21)) +
  theme_graph() 

ggsave(filename = "p2ry12_downstream_signal.pdf")
```


### number of DEGs in different order  
```{r}
tg <- as_tbl_graph(all_path_df)

DEG_14_orders <- tg %>% 
  dplyr::mutate(bfsr = bfs_dist(root = .N()$name == "70839", mode = "out")) %>% 
  dplyr::filter(.N()$name %in% intersect(mapIds(x = org.Mm.eg.db, keys = as.character(A14_DEG), column = "ENTREZID", keytype = "SYMBOL"), 
                                         mapIds(x = org.Mm.eg.db, keys = as.character(L14_DEG), column = "ENTREZID", keytype = "SYMBOL"))) %>% 
  as_tibble() %>% 
  pull(bfsr) %>% 
  table() %>% 
  as.data.frame() %>% 
  magrittr::set_colnames(c("order", 'freq'))

DEG_3_orders <- tg %>% 
  dplyr::mutate(dfsr = bfs_dist(root = .N()$name == "70839", mode = "out")) %>% 
  dplyr::filter(.N()$name %in% intersect(mapIds(org.Mm.eg.db, as.character(A3_DEG), "ENTREZID", "SYMBOL"), 
                                  mapIds(org.Mm.eg.db, as.character(L3_DEG), "ENTREZID", "SYMBOL"))) %>% 
  as_tibble() %>% 
  pull(dfsr) %>% 
  table() %>% 
  as.data.frame() %>% 
  magrittr::set_colnames(c("order", "freq"))

g %>% 
  dplyr::mutate(dfsr = bfs_dist(root = .N()$name == "P2ry12", mode = "out")) %>% 
  filter(.N()$name %in% all_DEG) %>% 
  as_tibble() %>% 
  filter(dfsr == 3) %>% 
  pull(name) -> o3_DEG
```


```{r}
DEG_3_orders$group = '3'
DEG_14_orders$group = '14'

DEG_3_orders %>% 
  rbind(DEG_14_orders) %>% 
  ggplot(aes(order, freq)) +
  geom_bar(stat = "identity", position = "dodge", aes(fill = group)) +
  scale_fill_brewer(name = "group", type = "qual", palette = "Set1", direction = -1) +
  ggtitle(label = "Number of DEGs in different order")

ggsave(filename = "Number-of-DEGs-in-different-order.tiff")
```


```{r}
all_DEG
```

```{r, warning=FALSE, message=FALSE}
tg %>% 
  dplyr::mutate(bfsr = bfs_dist(root = .N()$name == "70839", mode = "out")) %>% 
  dplyr::filter(.N()$name %in% intersect(mapIds(x = org.Mm.eg.db, keys = as.character(A14_DEG), column = "ENTREZID", keytype = "SYMBOL"), 
                                         mapIds(x = org.Mm.eg.db, keys = as.character(L14_DEG), column = "ENTREZID", keytype = "SYMBOL"))) %>% 
  as_tibble() %>% 
  mutate(symbol = mapIds(org.Mm.eg.db, keys = .$name, keytype = "ENTREZID", column = "SYMBOL")) %>% 
  magrittr::set_colnames(c("entrez", "order", "symbol")) %>% 
  write.csv(file = "A14_L14_genes_in_different_order.csv", row.names = F, quote = F)
```


```{r, warning=FALSE, message=FALSE}
tg %>% 
  dplyr::mutate(dfsr = bfs_dist(root = .N()$name == "70839", mode = "out")) %>% 
  dplyr::filter(.N()$name %in% intersect(mapIds(org.Mm.eg.db, as.character(A3_DEG), "ENTREZID", "SYMBOL"), 
                                  mapIds(org.Mm.eg.db, as.character(L3_DEG), "ENTREZID", "SYMBOL"))) %>% 
  as_tibble() %>% 
  mutate(symbol = mapIds(org.Mm.eg.db, keys = .$name, keytype = "ENTREZID", column = "SYMBOL")) %>% 
  magrittr::set_colnames(c("entrez", "order", "symbol")) %>% 
  write.csv(file = "A3_L3_genes_in_different_order.csv", row.names = F, quote = F)
```

# Go analysis  
```{r}
pvn_A3_DEG_up <- read.csv(file = "../data/20200702_3A_naive_DE_FDR.csv") %>% 
  dplyr::filter(logFC >= 1) %>% 
  select(geneid_symbol, logFC, PValue) %>% 
  dplyr::arrange(-logFC)

pvn_A3_DEG_down <- read.csv(file = "../data/20200702_3A_naive_DE_FDR.csv") %>% 
  dplyr::filter(logFC <= -1) %>% 
  select(geneid_symbol, logFC, PValue) %>% 
  dplyr::arrange(logFC)


pvn_L3_DEG_up <- read.csv(file = "../data/20200702_3L_naive_DE_FDR.csv") %>% 
  dplyr::filter(logFC >= 1) %>% 
  select(geneid_symbol, logFC, PValue) %>% 
  dplyr::arrange(-logFC)

pvn_L3_DEG_down <- read.csv(file = "../data/20200702_3L_naive_DE_FDR.csv") %>% 
  dplyr::filter(logFC <= -1) %>% 
  select(geneid_symbol, logFC, PValue) %>% 
  dplyr::arrange(logFC)


pvn_A14_DEG_up <- read.csv(file = "../data/20200702_14A_naive_DE_FDR.csv") %>% 
  dplyr::filter(logFC >= 1) %>% 
  select(geneid_symbol, logFC, PValue) %>% 
  dplyr::arrange(-logFC)

pvn_A14_DEG_down <- read.csv(file = "../data/20200702_14A_naive_DE_FDR.csv") %>% 
  dplyr::filter(logFC <= -1) %>% 
  select(geneid_symbol, logFC, PValue) %>% 
  dplyr::arrange(logFC)


pvn_L14_DEG_up <- read.csv(file = "../data/20200702_14L_naive_DE_FDR.csv") %>% 
  dplyr::filter(logFC >= 1) %>% 
  select(geneid_symbol, logFC, PValue) %>% 
  dplyr::arrange(-logFC)

pvn_L14_DEG_down <- read.csv(file = "../data/20200702_14L_naive_DE_FDR.csv") %>% 
  dplyr::filter(logFC <= -1) %>% 
  select(geneid_symbol, logFC, PValue) %>% 
  dplyr::arrange(logFC)
```


```{r}
pvn_A3_L3_DEG_up <- intersect(pvn_A3_DEG_up$geneid_symbol, pvn_L3_DEG_up$geneid_symbol)
pvn_A3_L3_DEG_down <- intersect(pvn_A3_DEG_down$geneid_symbol, pvn_L3_DEG_down$geneid_symbol)

pvn_A14_L14_DEG_up <- intersect(pvn_A14_DEG_up$geneid_symbol, pvn_L14_DEG_up$geneid_symbol)
pvn_A14_L14_DEG_down <- intersect(pvn_A14_DEG_down$geneid_symbol, pvn_L14_DEG_down$geneid_symbol)
```


```{r}
pvn_A3_L3_DEG_up_df <- select(x = org.Mm.eg.db, 
                              keys = pvn_A3_L3_DEG_up, 
                              keytype = "SYMBOL", 
                              columns = c("SYMBOL", "ENTREZID"))

pvn_A3_L3_DEG_down_df <- select(x = org.Mm.eg.db, 
                                keys = pvn_A3_L3_DEG_down, 
                                keytype = "SYMBOL", 
                                columns = c("ENTREZID", "SYMBOL"))


pvn_A14_L14_DEG_up_df <- select(x = org.Mm.eg.db, 
                                keys = pvn_A14_L14_DEG_up, 
                                keytype = "SYMBOL", 
                                columns = c("SYMBOL", "ENTREZID"))

pvn_A14_L14_DEG_down_df <- select(x = org.Mm.eg.db, 
                                  keys = pvn_A14_L14_DEG_down, 
                                  keytype = "SYMBOL", 
                                  columns = c("SYMBOL", "ENTREZID"))
```


```{r}
pvn_A3_L3_DEG_up_GO <- goana(de = pvn_A3_L3_DEG_up_df$ENTREZID, 
                             species = "Mm") %>% 
  tibble::rownames_to_column("GOID") %>% 
  mutate(group = "up")

pvn_A3_L3_DEG_down_GO <- goana(de = pvn_A3_L3_DEG_down_df$ENTREZID, 
                               species = "Mm") %>% 
  tibble::rownames_to_column("GOID") %>% 
  mutate(group = "down")

pvn_A3_L3_DEG_GO <- rbind(pvn_A3_L3_DEG_up_GO, pvn_A3_L3_DEG_down_GO)



pvn_A14_L14_DEG_up_GO <- goana(de = pvn_A14_L14_DEG_up_df$ENTREZID, 
                               species = "Mm") %>% 
  tibble::rownames_to_column("GOID") %>% 
  mutate(group = "up")

pvn_A14_L14_DEG_down_GO <- goana(de = pvn_A14_L14_DEG_down_df$ENTREZID, 
                                 species = "Mm") %>% 
  tibble::rownames_to_column("GOID") %>% 
  mutate(group = "down")

pvn_A14_L14_DEG_GO <- rbind(pvn_A14_L14_DEG_up_GO, pvn_A14_L14_DEG_down_GO)
```















