---
title: "normal AMD bulk sequencing"
author: "yincy"
date: "4/23/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# load package  
```{r, message=FALSE, warning=FALSE}
library(tidyverse)
library(edgeR)
library(limma)
library(biomaRt)
library(GEOquery)
library(EnsDb.Hsapiens.v86)
library(RColorBrewer)
library(pheatmap)
```


# get sample metadata  
```{r, cache=TRUE, cache.path="."}
GSE135092 <- getGEO(GEO = "GSE135092")
```


# load data  
```{r}
# geo accession of each group
amd_geo_accession <- GSE135092[[1]] %>% as.data.frame() %>% 
    dplyr::filter(tissue.ch1 == "RPE, Macula", amd_status.ch1 == "AMD") %>% 
    pull(geo_accession)

control_geo_accession <- GSE135092[[1]] %>% as.data.frame() %>% 
    dplyr::filter(tissue.ch1 == "RPE, Macula", amd_status.ch1 == "Control") %>% 
    pull(geo_accession)

# reads counts table for each file into an list
reads_list <- list()
for(i in c(amd_geo_accession, control_geo_accession)){
    reads = read.table(file = list.files(path = "/home/yincy/git/Data/AMD-related/AMD-bulk-seq/GSE135092/GSE135092/", 
                                                 pattern = i, 
                                                 full.names = T), 
                               header = T, 
                               stringsAsFactors = F,
                               sep = "\t")
    reads_list[[i]] <- reads
}


# check the equality of row names
a <- c()
for(i in seq_along(reads_list)){
    a[i] <- append(a, identical(reads_list[[1]][, 1], reads_list[[i]][, 1]))
}

all(a)


# from count list to data frame 
reads_table <- sapply(reads_list, FUN = function(x){cbind(x[, 3])}, simplify = T)
reads_table %>% dim()
reads_table %>% .[1:5, 1:5]
reads_table <- reads_table %>% 
    as.data.frame() %>% 
    magrittr::set_rownames(value = reads_list[[1]][, 1])

write.table(x = reads_table, 
            file = "/home/yincy/git/R-codes/Retina/bulk-amd-normal/reads_table_amd_ctrl.csv", 
            row.names = T, 
            col.names = T, 
            quote = F)

# chack the order 
identical(colnames(reads_table)[seq_along(amd_geo_accession)], amd_geo_accession)
identical(colnames(reads_table)[seq_along(c(amd_geo_accession, control_geo_accession))], c(amd_geo_accession, control_geo_accession))
```

```{r}
sample_info <- GSE135092[[1]] %>% as.data.frame() %>% 
    dplyr::filter(tissue.ch1 == "RPE, Macula") %>% 
    arrange(amd_status.ch1)

rownames(sample_info) <- c(amd_geo_accession, control_geo_accession)

g <- c(rep("amd", length(amd_geo_accession)), rep("ctrl", length(control_geo_accession)))

y <- DGEList(counts = reads_table, 
             samples = sample_info, 
             group = g)
```


## plot the sequencing depth    
```{r library size plot}
colSums(reads_table) %>% barplot(names.arg = NA)
abline(h = colSums(reads_table) %>% median(), col = "blue")
```


## filter out low expressed genes  
```{r calculate normalization factors}
y <- calcNormFactors(object = y, 
                     method = "TMM")
```

```{r plot normalized factors}
barplot(y$samples$norm.factors, col = "orange", border = NA, ylim = c(0, 1.3))
abline(h = y$samples$norm.factors %>% median(), col = "red", lty = 2, lwd = 1)
```


```{r cpm normalize}
cpm_reads <- edgeR::cpm(y = y, 
                normalized.lib.size = T)
```


```{r}
filter_thresh <- (10/(colSums(reads_table) %>% quantile(probs = 0.5))) * 1e6

keep <- rowSums(cpm_reads > filter_thresh) > length(y$samples$group) * 0.2
y_keep <- y[keep, ]
```


```{r}
plotMDS(y_keep, pch = c(1, 4)[y$samples$group], col = RColorBrewer::brewer.pal(n = 3, name = "Set1")[1:2][y$samples$group])

legend("topleft", legend = y$samples$group %>% unique(), 
       col = RColorBrewer::brewer.pal(n = 3, name = "Set1")[1:2][y$samples$group %>% unique()], 
       pch = c(1, 4)[y$samples$group %>% unique()])
```


```{r plot log noramlized cpm}
boxplot(log1p(cpm_reads[keep, ]),
        col = c("orange", "blue")[y$samples$group], 
        xaxt = "n", 
        main = "log1p cpm expression")

abline(h = log1p(cpm_reads[keep, ]) %>% median(), lty = 2, lwd = 1, col = "red")
```

```{r}
dist_tsne <- Rtsne::Rtsne(X = y_keep$counts %>% t(), 
                          theta = 0)
plot(dist_tsne$Y, 
     col = c("amd" = "red", "ctrl" = "blue")[y_keep$samples$group], 
     pch = 16)
legend("topright", 
       legend = y_keep$samples$group %>% unique(), 
       pch = 16, 
       col = c("amd" = "red", "ctrl" = "blue")[y_keep$samples$group %>% unique()])
```


# differential expression analysis  
```{r}
design <- model.matrix( ~ 0 + g)
colnames(design) <- gsub("g", "", colnames(design))
```


```{r}
v <- voom(counts = y_keep, design = design, plot = T)
```

```{r}
boxplot(v$E, xaxt = "n", col = c("orange", "blue")[y$samples$group])
abline(h = v$E %>% median(), lty = 2, lwd = 1, col = "red")
```

```{r}
var_geen_top500 <- apply(v$E, 1, var) %>% sort(decreasing = T) %>% head(n = 100) %>% names()
ann_col <- data.frame(
    state = rep(c("amd", "ctrl"), times = c(length(amd_geo_accession), length(control_geo_accession)))
)
rownames(ann_col) <- v$E %>% colnames()

pheatmap::pheatmap(mat = v$E[rownames(v$E) %in% ensem, ], 
                   annotation_col = ann_col, 
                   show_colnames = F, 
                   show_rownames = F, 
                   cluster_rows = T, 
                   clustering_method = "ward.D2", 
                   clustering_distance_cols = "correlation", 
                   clustering_distance_rows = "correlation")
```


```{r}
fit <- lmFit(object = v)
names(v)
```


```{r}
colnames(design)
cont_mat <- makeContrasts(AMD.vs.Control = amd - ctrl, levels = design)
fit_cont <- contrasts.fit(fit = fit, contrasts = cont_mat)
fit_cont <- eBayes(fit = fit_cont)
dim(fit_cont)
summ_fit <- decideTests(object = fit_cont, p.value = 0.1)
summary(summ_fit)
```

```{r}
deg_table <- topTable(fit = fit_cont, sort.by = "p", number = Inf)
write.table(x = deg_table, 
            file = "f:/git/R-codes/Retina/bulk-amd-normal/RPE_amd-control_deg_table.csv", 
            sep = "\t", 
            row.names = T, 
            quote = F, 
            col.names = T)
```


## KEGG and GO  
```{r}
degs <- deg_table %>% rownames_to_column(var = "ensembl") %>% 
    dplyr::filter(adj.P.Val <= 0.1) %>% 
    pull(ensembl)
```


```{r}
degs_genes <- select(x = EnsDb.Hsapiens.v86, 
       keys = degs, 
       column = c("GENEID", "ENTREZID", "SYMBOL"), 
       keytype = "GENEID")
```

```{r}
go <- goana(de = degs_genes$ENTREZID, species = "Hs")
go[order(go$P.DE), ]
write.table(x = go, file = "f:/git/R-codes/Retina/bulk-amd-normal/go_enrichment_table.csv", 
            sep = "\t", 
            row.names = T, 
            col.names = T)
```

```{r}
hmart <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
listAttributes(mart = hmart)
```

```{r}
searchFilters(hmart, "go")
```


```{r}
ensem <- getBM(attributes = c("entrezgene_id", "hgnc_symbol", "ensembl_gene_id"), 
      filters = "go", 
      values = c("GO:0031012"), 
      mart = hmart) %>% 
  pull(ensembl_gene_id) %>% 
  unique()
```



```{r}
ke <- kegga(de = degs_genes$ENTREZID)
ke[order(ke$P.DE), ]
write.table(x = ke, 
            file = "f:/git/R-codes/Retina/bulk-amd-normal/kegg_enrichment_table.csv",
            row.names = T, 
            col.names = T,
            sep = "\t")
```


```{r}
altered_paths <- ke[order(ke$P.DE), ] %>% head(n = 30) %>% rownames()
```


```{r}
source("f:/git/R-codes/my-functions/kegg-pathways.R")
```

```{r}
hsa_path <- KEGG_Pathways(database = "pathway", organism = "hsa")
hsa_path %>% names()
```


```{r}
path_name <- keggList("pathway", "hsa")
path_name[altered_paths]
```


## Metabolomics  
```{r}
metabolites <- read.csv(file = "f:/metabolites.csv", stringsAsFactors = F)
metabolites %>% 
    dplyr::filter(tolower(Sub.Pathway) == "tyrosine metabolism")
```

```{r}
intersect(x = metabolites %>% dplyr::filter(Qval.Sig == 1) %>% pull(Sub.Pathway) %>% tolower(), 
          y = ke[order(ke$P.DE), ] %>% dplyr::filter(P.DE <= 0.05) %>% pull(Pathway) %>% tolower())
```

## Tyrosine metabolism  
```{r}
tyrosine_metabolism_genes <- keggLink("hsa", "path:hsa00350")
tyrosine_metabolism_genes <- gsub("hsa:", "", tyrosine_metabolism_genes)
names(tyrosine_metabolism_genes) <- NULL
inter_genes <- intersect(tyrosine_metabolism_genes, degs_genes$ENTREZID)
inter_genes <- degs_genes[degs_genes$ENTREZID %in% inter_genes, ]
```

```{r}
inter_cpm <- cpm_reads[rownames(cpm_reads) %in% inter_genes$GENEID, ]

g_data <- apply(inter_cpm, 1, median)
names(g_data) <- inter_genes$ENTREZID

ann_col = data.frame(group = rep(c("Control", "AMD"), c(length(control_geo_accession), length(amd_geo_accession))))
rownames(ann_col) <- c(control_geo_accession, amd_geo_accession)

pheatmap::pheatmap(mat = inter_cpm, 
                   annotation_col = ann_col, 
                   clustering_method = "ward.D2", 
                   show_colnames = F, 
                   main = "variable genes in tyrosine metabolism")
```


```{r}
library(pathview)
pathview(gene.data = g_data, 
         cpd.data = c("C09368", "C01179"), 
         pathway.id = "hsa00350")
```


## gene set enrichment  
```{r}
library(fgsea)
library(qusage)
c2 <- read.gmt(file = "c:/Users/YinCY/Desktop/c2.all.v7.1.entrez.gmt")
```

```{r}
stats <- cpm_reads[rownames(cpm_reads) %in% degs, ]
stats <- apply(stats[, amd_geo_accession], 1, mean) / apply(stats[, control_geo_accession], 1, mean)
name_stats <- names(stats)

name_entrez <- c()
for(i in 1:dim(degs_genes)[1]){
    if(degs_genes[i, "GENEID"] %in% name_stats){
        name_entrez[i] <- degs_genes[i, "ENTREZID"]
    }
}

df <- as.data.frame(stats)
df <- rownames_to_column(df, "GENEID")

entrez <- select(EnsDb.Hsapiens.v86, 
       keys = names(stats), 
       keytype = "GENEID", 
       column = c("ENTREZID", "GENEID"))

df <- left_join(df, entrez, by = "GENEID")
df <- na.omit(df)
df <- df[!duplicated(df$ENTREZID), ]
df
rownames(df) <- df$ENTREZID
df <- df[, 1]
stats <- df[, 3]
names(stats) <- df$ENTREZID

amd_set <- fgsea(pathways = c2, 
                 stats = stats, 
                 nperm = 1000)
amd_set %>% 
    arrange(padj)
```



